---
title: "CRISPR-Cas Systems analysis"
author: "Thomas Nicholson"
date: "2/9/2017"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

##load packages
library(xtable)
library(gplots)
library(stringr)
library(ggplot2)
library(plyr)
library(devtools)
library(dplyr)
library(tidyr)
require(gridExtra)
require(cowplot)

##used to take a dataframe and assign correct data type (numeric, character, logical)
transform.df <- function(dat){
  ##loop that will go across the columns and look at each header
  for(i in 1:length(dat[1,])){
    
    ##checks if 'num' is in the header and assigns numeric if true
    if(substr(colnames(dat)[i],(nchar(colnames(dat)[i])-3),nchar(colnames(dat)[i]) )=='.num'){
      dat[,i] <- as.character(dat[,i])
      dat[,i] <- as.numeric(dat[,i])
      
    ##checks if 'y.n' is in the header and assigns logical if true
    }else if(substr(colnames(dat)[i], nchar(colnames(dat)[i])-3, nchar(colnames(dat)[i]))=='.y.n'){
      dat[,i] <- as.logical(dat[,i])

    ##assigns character to all the remaining columns
    }else{
      dat[,i] <- as.character(dat[,i])
    }
  }
  return(dat)
}

##count the number of times a subtype occurs in the genome list
subtypes.counts.funct <- function(subtype.list, genomes){
  subtype.count <- vector(mode = 'numeric', length = length(subtype.list))
  
  ##loop through the subtypes
  for(i in 1:length(subtype.list)){
    ##find all genomes containing a subtype
    x <- grep(subtype.list[i], genomes$subtypes)
    subtype.count[i] <- length(x)
  }
  subtypes.dat <- data.frame(subtype = subtype.list, counts = subtype.count)
  return(subtypes.dat)
  }
```

##Hmm output
Take the hmmsearch output and the file location output and generate a file that contains a list of all the putative cas proteins and their location on the server.
  - The file name needs to be updated to bacteria/archaea and the current Refseq version.

```{r file_locations, include=F}
##import the hmmsearch output
dat <- read.table("~/Desktop/Project/identifycasproteins/predicted_cas_genes_in_archaea_of_refseq83.txt", comment.char = "", sep = "", header = F, fill = T, col.names=c("target_name",        "accession",  "query_name",           "accession.2",    "E-value",  "score",  "bias",   "E-value.2",  "score.2",  "bias.2",   "exp", "reg", "clu" , "ov" ,"env" ,"dom" ,"rep", "inc" ,"description_of_target"))

##remove columns that are not hmm output (such as column names or comments)
dat <- dat%>%filter(description_of_target == "-")

##import the contig names and file locations
contig.lookup <- read.table("~/Desktop/Project/identifycasproteins/tmp.protein_name_lookup.txt", comment.char = "", fill = T)
colnames(contig.lookup) <- c("filepath", "target_name")

##select the target name from the hmmsearch output (contains the contig and protein ID)
proteinDat <- unique(dat%>%select(target_name))

##separate out the target name into each of the parts
proteinDat <- separate(proteinDat, target_name, c("contig.name", "protein.name", "refseq.description", "contig.length", "organism.name", "gene.start", "gene.end"), sep = "\\$", remove = T, extra = "merge")

##remove duplicate columns
contig.lookup <- unique(contig.lookup%>%select(filepath, target_name))

##separate out the target name into each of the parts
contig.lookup <- separate(contig.lookup, target_name, c("contig.name", "organism.name"), sep = "\\$", remove = F, extra = "merge")
contig.lookup <- contig.lookup%>%select(filepath, contig.name)

##combine protein name and file location by the contig name
proteinDat <- left_join(proteinDat, contig.lookup, by = "contig.name")
proteinDat <- proteinDat%>%select(filepath, protein.name)

##reformat the output so that extract_protein_sequences.py can work with it.
proteinDat <- proteinDat%>%group_by(filepath)%>%summarise(protein.list = paste(protein.name, collapse = ","))
proteinDat <- proteinDat%>%select(protein.list, filepath)

##append the current location of the Refseq database to the start of the file location and change the end from _genomic.gbff to _protein.faa
proteinDat <- proteinDat%>%mutate(filepath = paste("/DB/RefSeq83/archaea/", substr(as.character(filepath), 1, nchar(as.character(filepath)) - 13), "_protein.faa", sep = ""))
  
##write the data frame to a file
write.table(file = "~/Desktop/Project/identifycasproteins/archaea_refseq_83.filelocations.txt", x = proteinDat, quote = F, row.names = F, col.names = F, sep = "\t")
```


## Hmm and rps output
Take the hmmsearch results and the rps search results. These will be combined and then analysed to identify the cas proteins in genomes
  - The file name needs to be updated to bacteria/archaea and the current Refseq version.

Included in the upload is a file containing all of the proteins with CRISPR or Cas in the description which will be used to ensure no cas protein is missed.
Cas gene assigned if:
  - Protein description describes protein as Cas
  - Cas model scores better than generic model
  - Labelled as putative if the Cas model scores close to the rps model

Analysis is carried out on the putative cas genes
  - Check that the function in the description is the function of the cas gene
  - Check if the protein has a generic or hypothetical description

The genes assigned cas are then combined into the genomes that they are from. The signature genes are also combined to allow the genome subtypes to be assigned.
The number of proteins and the number of subtypes is counted.

Genomes with a single subtype are written to a file (This is currently what is done but I might change this approach)


```{r cas_gene_assign}
##Import data
##import the hmmsearch output
hmmB <- read.table("~/Desktop/Project/identifycasproteins/predicted_cas_genes_all_hash.txt", comment.char = "", sep = "", header = F, fill = T)
hmmA <- read.table("~/Desktop/Project/identifycasproteins/predicted_cas_genes_in_archaea_of_refseq83.txt", comment.char = "", sep = "", header = F, fill = T, col.names=c("target_name",        "accession",  "query_name",           "accession.2",    "E-value",  "score",  "bias",   "E-value.2",  "score.2",  "bias.2",   "exp", "reg", "clu" , "ov" ,"env" ,"dom" ,"rep", "inc" ,"description_of_target"))

##remove columns that are not hmm output (such as column names or comments)
hmmB <- hmmB%>%filter(V19 == "-")
hmmB <- hmmB[,1:19]

##add column names
colnames(hmmB) <- c("target_name", "accession","query_name","accession.2","E.value","score","bias","E.value.2","score.2","bias.2","exp","reg","clu","ov","env","dom","rep","inc","description_of_target")                  

##remove columns that are not hmm output (such as column names or comments)
hmmA <- hmmA%>%filter(description_of_target == "-")

##combine archaea and bacteria
hmm <- rbind(hmmA, hmmB)


##import rps results
rpsA <- read.table("~/Desktop/Project/identifycasproteins/archaea_refseq_83.cdd.out", comment.char = "#", sep = "", header = F, fill = T, col.names = c( "query_acc.ver", "subject_acc.ver", "identity_%", "alignment_length", "mismatches", "gap_opens", "q.start", "q.end", "s.start", "s.end", "evalue", "bit_score"))
rpsB <- read.table("~/Desktop/Project/identifycasproteins/bacteria_refseq_83.cdd.out", comment.char = "#", sep = "", header = F, fill = T, col.names = c( "query_acc.ver", "subject_acc.ver", "identity_%", "alignment_length", "mismatches", "gap_opens", "q.start", "q.end", "s.start", "s.end", "evalue", "bit_score"))

##combine archaea and bacteria
rps <- rbind(rpsA, rpsB)

##add column names
colnames(rps) <- c("Protein_ID", "subject_acc.ver","identity_.","alignment_length","mismatches.rps","gap_opens.rps" ,"q.start","q.end","s.start","s.end","evalue","bit_score")

##import the cas genes selected from the descriptions
descr <- read.table("~/Desktop/Project/identifycasproteins/cas_genes_descriptions.txt", comment.char = "", sep = "", header = F, fill = T)

##split out information into relevant fields
descr <- separate(descr, V1, into = c("contig", "Protein_ID", "Protein.Description", "contig.length", "Genome", "gene.start", "gene.stop"), sep = "\\$")

##change contig column so the names match the other files
descr <- descr%>%mutate(contig = substr(as.character(contig), 2, nchar(as.character(contig))))

##import the contig names and file locations
contig.lookupA <- read.table("~/Desktop/Project/identifycasproteins/archaea.protein_name_lookup.txt", comment.char = "", fill = T)
contig.lookupB <- read.table("~/Desktop/Project/identifycasproteins/bacteria.protein_name_lookup.txt", comment.char = "", fill = T)

##add column names
colnames(contig.lookupA) <- c("filepath", "target_name")
colnames(contig.lookupB) <- c("filepath", "target_name")

##remove duplicates
contig.lookupA <- unique(contig.lookupA)
contig.lookupB <- unique(contig.lookupB)

##add column with info used for file location
contig.lookupA <- contig.lookupA%>%mutate(domain.group = "Archaea")
contig.lookupB <- contig.lookupB%>%mutate(domain.group = "Bacteria")

##combine archaea and bacteria
contig.lookup <- rbind(contig.lookupB, contig.lookupA)


##remove duplicate rows
hmm <- unique(hmm)

##split out information into relevant fields
hmm <- separate(hmm, target_name, into = c("contig", "Protein_ID", "Protein.Description", "contig.length", "Genome", "gene.start", "gene.stop"), sep = "\\$", remove = F)

##make column with unique ID for each protein 
hmmT <- hmm%>%mutate(protein.genome = paste(Protein_ID, contig, sep = "$"))%>%select(protein.genome)%>%mutate(cas.group = "Cas genes")
descrT <- descr%>%mutate(protein.genome = paste(Protein_ID, contig, sep = "$"))%>%select(protein.genome)%>%mutate(descr.group = "Cas descriptions")

##combine selection from description and from HMMs
hmmD <- left_join(hmmT, descrT, by = "protein.genome")
descrH <- left_join(descrT, hmmT, by = "protein.genome")

##reorder columns
descrH <- descrH%>%select(protein.genome, cas.group, descr.group)

##combine lists for full set of proteins
proteins <- rbind(hmmD, descrH)

##remove duplicate rows
proteins <- unique(proteins)

##label the proteins based on whether they were new, missed or found in both sets
proteins <- proteins%>%
  mutate(cas.group = ifelse(is.na(cas.group), "new", cas.group))%>%
  mutate(descr.group = ifelse(is.na(descr.group), "not found", descr.group))%>%
  mutate(group = ifelse(cas.group == "Cas genes", ifelse(descr.group == "Cas descriptions", "-", "new"), ifelse(descr.group == "Cas descriptions", "Not found", "")))%>%
  select(-cas.group, -descr.group)

##change E.value to numeric. Select the smallest evaule for each protein.
hmmSmall <- hmm%>%mutate(E.value = as.numeric(as.character(E.value)))%>%group_by(Protein_ID)%>%mutate(x = min(E.value))%>%filter(E.value == x)

##change evalue to numeric. Select the smallest evaule for each protein.
rpsSmall <- rps%>%mutate(evalue = as.numeric(as.character(evalue)))%>%group_by(Protein_ID)%>%mutate(x = min(evalue))%>%filter(evalue == x)

##combine hmm and rps data
dat <- left_join(hmmSmall, rpsSmall, by = "Protein_ID")

##determine if a protein is a Cas protein by looking at the description for CRISPR or Cas and checking if the hmm result was better than the rps result.
dat <- dat%>%select(contig, Protein_ID, Protein.Description, gene.start, gene.stop, query_name, E.value, evalue, subject_acc.ver)%>%
  mutate(cas.gene = ifelse(grepl("CRISPR", Protein.Description), "Cas", ifelse(grepl("Cas", Protein.Description), "Cas", ifelse(E.value < evalue, "Cas", ifelse(is.na(evalue), "Cas", "-")))))
dat <- dat%>%mutate(cas.gene = ifelse(is.na(evalue), "Cas", cas.gene))

##label gene as putative if the hmm match was high and the rps match was not a lot higher.
dat <- dat%>%mutate(cas.gene = ifelse(cas.gene != "Cas", ifelse(evalue > 1e-50, ifelse(E.value < 1e-30, "Putative", cas.gene), cas.gene), cas.gene))

##Reformat descriptions data set so that it matches the dat dataset
dSmall <- descr%>%select(-contig.length, -Genome)%>%mutate(query_name = rep(NA, nrow(descr)))%>%mutate(E.value = rep(NA, nrow(descr)))%>%mutate(evalue = rep(NA, nrow(descr)))%>%mutate(subject_acc.ver = rep(NA, nrow(descr)))%>%mutate(cas.gene = rep("Cas", nrow(descr)))

dat <- as.data.frame(dat)

##combine hmm/rps results and descriptions
dat <- rbind(dat, dSmall)

##make column with unique ID for each protein 
dat <- dat%>%mutate(protein.genome = paste(Protein_ID, contig, sep = "$"))

##combine dat data and the info on whether the protein had previously been classified as Cas.
tt <- left_join(dat, proteins, by = "protein.genome")

##remove descriptions rows that are also found in hmm/rps as these have more info.
tt <- tt%>%mutate(keep = ifelse(is.na(E.value), ifelse(group == "Not found", "y", "n"), "y"))
tt <- tt%>%filter(keep == "y")

##import cas models data
signatures <- read.table("~/Desktop/Project/identifycasproteins/models_and_systems.txt", header = T, sep = '\t', comment.char = '', as.is = T, fill = T)

##relabel a column
colnames(tt)[6] <- "Profile_ID"

##combine the cas models and protien data
tt <- left_join(tt, signatures, by = "Profile_ID")

##data frame with the other data
casDat <- tt%>%filter(group != "new")

##data frame with new proteins
newDat <- tt%>%filter(group == "new")%>%filter(cas.gene != "-")

##data frame with putative genes only 
putativeDat <- newDat%>%filter(cas.gene == "Putative")

##get all the gene names
gene.list <- unique(tt%>%filter(!is.na(Gene_name))%>%select(Gene_name))
gene.list <- gene.list$Gene_name

##add column with lower cas gene descriptions
putativeDat <- putativeDat%>%mutate(Protein.Description.lower = tolower(Protein.Description))
putativeDat <- putativeDat%>%mutate(matching.description = rep("", nrow(putativeDat)))
i <- "cas4"

##check that each gene call matches the gene description
for(i in gene.list){
  print(i)
putativeDat <- putativeDat%>%mutate(matching.description = ifelse(grepl(i, Protein.Description.lower), i, matching.description))
}


##determine if the protein assignment matches the description
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(matching.description != "", ifelse(Gene_name == matching.description, "y", "n"), ""))


##check for correct function or similar gene names
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas3", Gene_name), ifelse(grepl("cas3", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas5", Gene_name), ifelse(grepl("cas5", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas6", Gene_name), ifelse(grepl("cas6", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas7", Gene_name), ifelse(grepl("cas7", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas8", Gene_name), ifelse(grepl("cas8", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas6", Gene_name), ifelse(grepl("cas6", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("WYL", Gene_name), ifelse(grepl("wyl", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas3", Gene_name), ifelse(grepl("helicase", Protein.Description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cas3", Gene_name), ifelse(grepl("nuclease", Protein.Description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("csa3", Gene_name), ifelse(grepl("transcription", Protein.Description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("csa3", Gene_name), ifelse(grepl("DNA-binding", Protein.Description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("csa3", Gene_name), ifelse(grepl("dna-binding", Protein.Description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("DinG", Gene_name), ifelse(grepl("helicase", Protein.Description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("csc2", Gene_name), ifelse(grepl("cas7", matching.description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("csc1", Gene_name), ifelse(grepl("cas5", matching.description), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("hypothetical", Protein.Description.lower), "y", correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cmr1", Gene_name), ifelse(grepl("cmr1", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cmr4", Gene_name), ifelse(grepl("cmr4", Protein.Description.lower), "y", correct.call), correct.call))
putativeDat <- putativeDat%>%mutate(correct.call = ifelse(grepl("cmr6", Gene_name), ifelse(grepl("cmr6", Protein.Description.lower), "y", correct.call), correct.call))

 
##keep correctly assigned genes
putativeDat <- putativeDat%>%filter(correct.call == "y")%>%select(-correct.call, -Protein.Description.lower, - matching.description)

##combine data
newDat <- newDat%>%filter(cas.gene != "Putative")
newDat <- rbind(newDat, putativeDat)
casDat <- casDat%>%filter(cas.gene != "-")
casDat <- rbind(casDat, newDat)


##split out information into relevant fields
contig.lookup <- separate(contig.lookup, target_name, into = c("contig", "Genome"), sep = "\\$", remove = F)


##combine contig data and cas data
casDat <- left_join(casDat, contig.lookup, by = "contig")

newDat <- casDat%>%filter(is.na(Gene_name))

##get all the gene names
gene.list <- unique(casDat%>%filter(!is.na(Gene_name))%>%select(Gene_name))
gene.list <- gene.list$Gene_name

##add column with lower cas gene descriptions
newDat <- newDat%>%mutate(Protein.Description.lower = tolower(Protein.Description))
i <- "cas4"

##check that each gene call matches the gene description
for(i in gene.list){
  print(i)
newDat <- newDat%>%mutate(Gene_name = ifelse(grepl(i, Protein.Description.lower), i, Gene_name))
}


casDat <- casDat%>%filter(!is.na(Gene_name))
newDat <- newDat%>%select(-Protein.Description.lower)

casDat <- rbind(casDat, newDat)

casDat <- casDat%>%mutate(Gene_name = ifelse(is.na(Gene_name), ifelse(grepl("Cse2", Protein.Description), "Cse2", Gene_name), Gene_name))

casDat <- casDat%>%mutate(Gene_name = ifelse(is.na(Gene_name), ifelse(grepl("Cse1", Protein.Description), "Cse1", Gene_name), Gene_name))
casDat <- casDat%>%mutate(keep.not_found = ifelse(is.na(Gene_name), ifelse(grepl("case", Protein.Description), F, T), T))%>%filter(keep.not_found)%>%select(-keep.not_found)

casDat <- casDat%>%mutate(Gene_name = ifelse(Gene_name == "cas1", "cas01", Gene_name))

nrow(unique(casDat%>%filter(domain.group == "Bacteria")%>%select(filepath)))
nrow(unique(casDat%>%filter(domain.group == "Archaea")%>%select(filepath)))

nrow(unique(contig.lookup%>%filter(domain.group == "Archaea")%>%select(filepath)))
nrow(unique(contig.lookup%>%filter(domain.group == "Bacteria")%>%select(filepath)))


tt <- casDat%>%
  filter(Gene_name != "DinG")%>%
  filter(Gene_name != "DEDDh")%>%
  filter(Gene_name != "WYL")%>%
  filter(Gene_name != "RT")%>%
  filter(Gene_name != "PD-DExK")%>%
  filter(Gene_name != "PrimPol")

nrow(unique(tt%>%filter(domain.group == "Bacteria")%>%select(filepath)))
nrow(unique(tt%>%filter(domain.group == "Archaea")%>%select(filepath)))

tt <- rbind(casDat%>%filter(grepl("cas01", Gene_name)), casDat%>%filter(grepl("cas2", Gene_name)), casDat%>%filter(grepl("cas3", Gene_name)), casDat%>%filter(grepl("cas4", Gene_name)), casDat%>%filter(grepl("cas6", Gene_name)), casDat%>%filter(grepl("cas8", Gene_name)), casDat%>%filter(grepl("cas9", Gene_name)), casDat%>%filter(grepl("cas5", Gene_name)), casDat%>%filter(grepl("cas7", Gene_name)))
nrow(unique(tt%>%filter(domain.group == "Bacteria")%>%select(filepath)))
nrow(unique(tt%>%filter(domain.group == "Archaea")%>%select(filepath)))


tt <- casDat%>%arrange(gene.start)%>%group_by(filepath)%>%summarise(cas.proteins = paste(unique(Gene_name), collapse = ", "))
protein.counts <- casDat%>%arrange(gene.start)%>%group_by(filepath)%>%summarise(protein.count = length(unique(Gene_name)))#%>%filter(protein.count > 2)
protein.counts <- left_join(protein.counts, tt, by = "filepath")
protein.counts <- protein.counts%>%mutate(cas1 = grepl("cas01,", cas.proteins))%>%mutate(cas2 = grepl("cas2,", cas.proteins))%>%mutate(basic.genes = ifelse(cas1, ifelse(cas2, T, F), F))

subtypes <- casDat%>%filter(!is.na(x))%>%filter(x != "-")%>%filter(x != "I-U")%>%arrange(gene.start)%>%group_by(filepath)%>%summarise(subtype.list = paste(unique(x), collapse = "_"))
subtype.counts <- casDat%>%filter(!is.na(x))%>%filter(x != "-")%>%filter(x != "I-U")%>%arrange(gene.start)%>%group_by(filepath)%>%summarise(subtype.count = length(unique(x)))


protein.counts <- left_join(protein.counts, subtypes, by = "filepath")
protein.counts <- left_join(protein.counts, subtype.counts, by = "filepath")

protein.counts <- protein.counts%>%mutate(type.ii = grepl("cas9", cas.proteins))



iiDat <- protein.counts%>%filter(type.ii)%>%mutate(ii.a = grepl("csn2", cas.proteins))%>%mutate(ii.b = grepl("cas4", cas.proteins))%>%mutate(ii.c = ifelse(ii.a == F, ifelse(ii.b == F, T, F), F))

iiDat <- left_join(casDat, iiDat, by = "filepath")

iiDat <- iiDat%>%filter(type.ii)%>%select(-subtype.list, - subtype.count)

iiDat <- iiDat%>%mutate(x = ifelse(Gene_name == "cas9", ifelse(ii.a,ifelse(ii.b, "II-B", ""), ifelse(ii.b, "II-B", "II-C")), x))

iiSub <- iiDat%>%filter(!is.na(x))%>%filter(x != "-")%>%filter(x != "I-U")%>%group_by(filepath)%>%summarise(tmp = paste(unique(sort(x)), collapse = "_"))%>%mutate(tmp = ifelse(substr(tmp, 1, 1) == "_", substr(tmp, 2, nchar(tmp)), tmp))%>%mutate(tmp = ifelse(substr(tmp, nchar(tmp), nchar(tmp)) == "_", substr(tmp, 1, nchar(tmp) - 1), tmp))

subDat <- protein.counts#%>%filter(!is.na(subtype.list))

subDat <- left_join(subDat, iiSub, by = "filepath")

subDat <- subDat%>%mutate(subtype.list = ifelse(!is.na(tmp), tmp, subtype.list))%>%select(-type.ii, -tmp)

subDat <- subDat%>%mutate(subtype.count = str_count(subtype.list, "_") + 1)
subDat <- subDat%>%mutate(single.system = ifelse(subtype.count == 1, T, ifelse(subtype.count == 2, ifelse(grepl("V", subtype.list), ifelse(grepl("IV", subtype.list), F, T), F), F)))


table(subDat$subtype.count)
table(subDat$single.system)

singleDat <- subDat%>%filter(single.system)#%>%filter(protein.count > 1)


singleDat <- singleDat%>%mutate(subtype = subtype.list)%>%select(filepath, subtype)
fileDat <- unique(contig.lookup%>%select(filepath, domain.group))

singleDat <- left_join(singleDat, fileDat, by = "filepath")

subDat <- separate(subDat, filepath, into = c("i1", "i2"), sep = "#G", remove = F, extra = "merge")
subDat <- separate(subDat, i2, into = "assembly_accession", sep = "/", remove = T, extra = "drop")

subDat<- subDat%>%mutate(assembly_accession = paste("G", assembly_accession, sep = ""))%>%select(-i1)
subDat <- subDat%>%mutate(cas1 = ifelse(grepl("cas01", cas.proteins), T, F))%>%mutate(cas2 = ifelse(grepl("cas2", cas.proteins), T, F))%>%mutate(basic.genes = ifelse(cas1, ifelse(cas2, T, F), F))

write.table(x = subDat, file = "~/Desktop/Project/identifycasproteins/cas_genomes_refseq_83.txt", quote = F, col.names = T, row.names = F, sep = "\t")
write.table(x = casDat, file = "~/Desktop/Project/identifycasproteins/cas_genes_refseq_83.txt", quote = F, col.names = T, row.names = F, sep = "\t")
write.table(x = singleDat, file = "~/Desktop/Project/identifycasproteins/cas_genomes_single_system_refseq_83.txt", quote = F, col.names = T, row.names = F, sep = "\t")


```



All of the CRISPRDetect files from bacteria_redo_v7 and archaea_redo were combined into a single file. These were then reformatted into a fasta format to allow the swipe nucleotide search to be done. 

The fasta formatted version is read into r along with the assembly summary data.

```{r CRISPRDetect}
assemB <- read.table("~/Desktop/Project/data_analysis_crispr/assembly_summary_bacteria_83.txt", header = F, sep = '\t', comment.char = '#', as.is = T, fill = T, quote = "")
assemA <- read.table("~/Desktop/Project/data_analysis_crispr/assembly_summary._archaea_83.txt", header = T, sep = '\t', comment.char = '#', as.is = T, fill = T, quote = "")
colnames(assemB) <- colnames(assemA)
assemA <- assemA%>%mutate(domainName = "Archaea")
assemB <- assemB%>%mutate(domainName = "Bacteria")
assem <- rbind(assemA, assemB)

subDat <- read.table( "~/Desktop/Project/identifycasproteins/cas_genomes_refseq_83.txt", quote = "", comment.char = "", header = T, fill = T, as.is = T, sep = "\t")

assem <- left_join(assem, subDat, by = "assembly_accession")
assem <- assem%>%mutate(basic.genes = ifelse(is.na(basic.genes), F, basic.genes))%>%mutate(protein.count = ifelse(is.na(protein.count), 0, protein.count))


spacers <- read.table("~/Desktop/Project/identifycasproteins/all_CRISPRDetect_refseq_83.fna", fill = T, comment.char = "#", quote = "")

spacers <- spacers%>%filter(grepl(">", V1))

##separate out the information from the spacer id
mat <- separate(spacers, V1,c("i1","i2","i3"), sep = "\\|", remove = F)
mat <- separate(mat, i1,c("contig","Organism"), sep = "-", remove = T)

mat <- mat%>%filter(grepl("#", contig)  == F)

mat <- separate(mat, i2,c("array.start.num","array.stop.num"), sep = "-", remove = T)


#mat <- separate(mat, Subtype,c("Type"), sep = "-", remove = F,  extra = "drop")
mat <- separate(mat, i3,c("i5", "i6"), sep = "\\[", remove = T, extra = "drop")
mat <- separate(mat, i5,c("array.num","spacer.num"), sep = "_", remove = T, extra = "merge")
mat <- separate(mat, i6,c("left.flank.num","right.flank.num", "array.score.num", "Taxonomy"), sep = "_", remove = T, extra = "drop")


spacers <- mat%>%mutate(contig = substr(contig, 2, nchar(as.character(contig))))

contigDat <- unique(contig.lookup%>%select(-target_name))



spacers <- full_join(spacers, contigDat, by = "contig")

tt <- spacers%>%filter(!is.na(array.score.num))%>%select(contig, filepath)
tt <- unique(tt)


write.table(x = tt, file = "~/Desktop/Project/identifycasproteins/cas_arrays_genomes_refseq_83.txt", quote = F, col.names = T, row.names = F, sep = "\t")

tt <- read.table("~/Desktop/Project/identifycasproteins/cas_arrays_genomes_refseq_83.txt", fill = T, comment.char = "#", quote = "", header = T)

spacerLoc <- read.table("~/Desktop/Project/identifycasproteins/CRISPRDetect_file_locations_refseq_83.txt", fill = T, comment.char = "", quote = "", header = F)

spacerLoc <- separate(spacerLoc, V1, into = c("i1","mnt", "SSD", "CD_Tracr_Cas9", "domainName", "letter", "i2", "i3"), sep = "/", remove = F)
spacerLoc <- separate(spacerLoc, i2, into = c("Genome", "assembly_accession"), sep = "\\#", extra = "merge")
spacerLoc <- separate(spacerLoc, assembly_accession, into = c("i4", "i5"), sep = "\\#", extra = "merge", remove = F)

spacerLoc <- spacerLoc%>%mutate(assembly_accession = ifelse(is.na(i5), assembly_accession, i5))%>%select(-i1, -i4, -i5, -mnt, -SSD, -CD_Tracr_Cas9)

spacerLoc <- spacerLoc%>%mutate(contig = substr(i3, 1, nchar(as.character(i3)) - 11))%>%select(-i3)

spacerLoc <- spacerLoc%>%mutate(domainName = ifelse(grepl("bacteria", domainName), "Bacteria", "Archaea"))


spacerGenomes <- spacerLoc%>%group_by(assembly_accession)%>%summarise(contig.count = n())

assem <- left_join(assem, spacerGenomes, by = "assembly_accession")

assem <- assem%>%mutate(proteins.present = ifelse(!is.na(cas.proteins), T, F))%>%mutate(arrays.present = ifelse(!is.na(contig.count), T, F))%>%mutate(subtype.present = ifelse(!is.na(subtype.count), T, F))%>%mutate(array.and_subtype = ifelse(subtype.present, ifelse(arrays.present, "True System", "No array"), ifelse(arrays.present, "Array only", "No System")))


tt <- assem

assem <- assem%>%filter(grepl("GCF", assembly_accession))%>%
  filter(grepl("ftp:", ftp_path))%>%
  mutate(wgs_master = ifelse(wgs_master == "", "-", wgs_master))%>%
  mutate(refseq_category = ifelse(refseq_category == "na", "-", refseq_category))%>%
  mutate(infraspecific_name = ifelse(infraspecific_name == "", "-", infraspecific_name))%>%
  mutate(isolate = ifelse(isolate == "", "-", isolate))%>%
  mutate(excluded_from_refseq = ifelse(excluded_from_refseq == "", "-", excluded_from_refseq))%>%
  mutate(excluded_from_refseq = ifelse(is.na(excluded_from_refseq), "-", excluded_from_refseq))%>%
  mutate(relation_to_type_material = ifelse(relation_to_type_material == "", "-", relation_to_type_material))%>%
  mutate(filepath = ifelse(is.na(filepath), "-", filepath))%>%
  mutate(cas.proteins = ifelse(is.na(cas.proteins), "-", cas.proteins))%>%
  mutate(cas1 = ifelse(is.na(cas1), F, cas1))%>%
  mutate(cas2 = ifelse(is.na(cas2), F, cas2))%>%
  mutate(basic.genes = ifelse(is.na(basic.genes), F, basic.genes))%>%
  mutate(subtype.list = ifelse(is.na(subtype.list), "-", subtype.list))%>%
  mutate(subtype.count = ifelse(is.na(subtype.count), 0, subtype.count))%>%
  mutate(single.system = ifelse(is.na(single.system), F, single.system))%>%
  mutate(contig.count = ifelse(is.na(contig.count), F, contig.count))

write.table(assem, file = "~/Desktop/Project/identifycasproteins/assembly_summary_refseq83_all.txt", quote = F, col.names = T, row.names = F, sep = "\t")


spacerLoc <- left_join(spacerLoc, subDat, by = "assembly_accession")

spacerSubtypes <- spacerLoc%>%select(contig, assembly_accession, subtype.list, single.system, domainName)

spacerSubtypes <- unique(spacerSubtypes)

```




```{r assembly_stuff}

assembySummary <- read.table("~/Desktop/Project/identifycasproteins/assembly_summary_refseq83_all.txt", header = T, sep = '\t', comment.char = "", as.is = T, fill = T, quote = "")
assembySummary <- assembySummary%>%filter(protein.count >= 0)

spacersFna <- read.table("~/Desktop/Project/identifycasproteins/all_CRISPRDetect_refseq_83.fna", fill = T, comment.char = "#", quote = "")
spacersFna <- spacersFna%>%filter(V2 != "")

##separate out the information from the spacer id
mat <- separate(spacersFna, V1,c("i1","i2","i3"), sep = "\\|", remove = F)
mat <- separate(mat, i1,c("contig","Organism"), sep = "-", remove = T, extra = "merge")
mat <- mat%>%filter(grepl("#", contig)  == F)
mat <- separate(mat, i2,c("array.start.num","array.stop.num"), sep = "-", remove = T)
mat <- separate(mat, i3,c("i5", "i6"), sep = "\\[", remove = T, extra = "drop")
mat <- separate(mat, i5,c("array.num","spacer.num"), sep = "_", remove = T, extra = "merge")
mat <- separate(mat, i6,c("left.flank.num","right.flank.num", "array.score.num", "Taxonomy"), sep = "_", remove = T, extra = "drop")
spacersFna <- mat%>%mutate(contig = substr(contig, 2, nchar(as.character(contig))))


spacerLocations <- read.table("~/Desktop/Project/identifycasproteins/CRISPRDetect_file_locations_refseq_83.txt", fill = T, comment.char = "", quote = "", header = F)
spacerLocations <- separate(spacerLocations, V1, into = c("i1","mnt", "SSD", "CD_Tracr_Cas9", "domainName", "letter", "i2", "i3"), sep = "/", remove = F)
spacerLocations <- separate(spacerLocations, i2, into = c("Genome", "assembly_accession"), sep = "\\#", extra = "merge")
spacerLocations <- separate(spacerLocations, assembly_accession, into = c("i4", "i5"), sep = "\\#", extra = "merge", remove = F)
spacerLocations <- spacerLocations%>%mutate(assembly_accession = ifelse(is.na(i5), assembly_accession, i5))%>%select(-i1, -i4, -i5, -mnt, -SSD, -CD_Tracr_Cas9)
spacerLocations <- spacerLocations%>%mutate(contig = substr(i3, 1, nchar(as.character(i3)) - 11))%>%select(-i3)
spacerLocations <- spacerLocations%>%mutate(domainName = ifelse(grepl("bacteria", domainName), "Bacteria", "Archaea"))

spacersFna <- left_join(spacersFna, spacerLocations, "contig")

spacerSubtypes <- left_join(spacersFna, assembySummary%>%select(assembly_accession, subtype.list, single.system)%>%distinct(), by = "assembly_accession")

spacerSubtypes <- spacerSubtypes%>%filter(single.system)%>%select(V1.x, V2)





array.counts <- spacersFna%>%filter(spacer.num == 1)%>%group_by(assembly_accession)%>%summarise(n.of.arrays = n())
spacer.counts <- spacersFna%>%group_by(assembly_accession)%>%summarise(n.of.spacers = n())


assembySummary <- left_join(assembySummary, array.counts, by = "assembly_accession")
assembySummary <- left_join(assembySummary, spacer.counts, by = "assembly_accession")



####functions here

assem <- assembySummary

assemRep <- assem%>%filter(refseq_category == "representative genome")
assemComplete <- assem%>%filter(assembly_level == "Complete Genome")
assemContig <- assem%>%filter(assembly_level == "Contig")

assemA <- assem%>%filter(domainName == "Archaea")
assemB <- assem%>%filter(domainName == "Bacteria")

casSummary <- as.data.frame(table(assem$domainName))
tmp <- assem%>%filter(refseq_category == "representative genome")
casSummary <- rbind(casSummary, as.data.frame(table(tmp$refseq_category)))
tmp <- assem%>%filter(assembly_level == "Complete Genome")
casSummary <- rbind(casSummary, as.data.frame(table(tmp$assembly_level)))
casSummary <- rbind(casSummary, as.data.frame(table(assemContig$assembly_level)))

assemAR <- assemA%>%filter(refseq_category == "representative genome")%>%mutate(refseq_category = "archaeal representative genome")
casSummary <- rbind(casSummary, as.data.frame(table(assemAR$refseq_category)))

assemBR <- assemB%>%filter(refseq_category == "representative genome")%>%mutate(refseq_category = "bacterial representative genome")
casSummary <- rbind(casSummary, as.data.frame(table(assemBR$refseq_category)))


assemAC <- assemA%>%filter(assembly_level == "Complete Genome")%>%mutate(assembly_level = "Archaeal Complete Genome")
casSummary <- rbind(casSummary, as.data.frame(table(assemAC$assembly_level)))

assemBC <- assemB%>%filter(assembly_level == "Complete Genome")%>%mutate(assembly_level = "Bacterial Complete Genome")
casSummary <- rbind(casSummary, as.data.frame(table(assemBC$assembly_level)))


colnames(casSummary) <- c("Category", "Genome.count")





casSummary <- casSummary%>%mutate(Genomes.with.cas.proteins = c(as.data.frame(table(assemA$proteins.present))[2,2],
as.data.frame(table(assemB$proteins.present))[2,2],
as.data.frame(table(assemRep$proteins.present))[2,2],
as.data.frame(table(assemComplete$proteins.present))[2,2],
as.data.frame(table(assemContig$proteins.present))[2,2],
as.data.frame(table(assemAR$proteins.present))[2,2],
as.data.frame(table(assemBR$proteins.present))[2,2],
as.data.frame(table(assemAC$proteins.present))[2,2],
as.data.frame(table(assemBC$proteins.present))[2,2]))


casSummary <- casSummary%>%mutate(Genomes.with.arrays = c(as.data.frame(table(assemA$arrays.present))[2,2],
as.data.frame(table(assemB$arrays.present))[2,2],
as.data.frame(table(assemRep$arrays.present))[2,2],
as.data.frame(table(assemComplete$arrays.present))[2,2],
as.data.frame(table(assemContig$arrays.present))[2,2],
as.data.frame(table(assemAR$arrays.present))[2,2],
as.data.frame(table(assemBR$arrays.present))[2,2],
as.data.frame(table(assemAC$arrays.present))[2,2],
as.data.frame(table(assemBC$arrays.present))[2,2]))

casSummary <- casSummary%>%mutate(Genomes.with.subtypes.identified = c(as.data.frame(table(assemA$subtype.present))[2,2],
as.data.frame(table(assemB$subtype.present))[2,2],
as.data.frame(table(assemRep$subtype.present))[2,2],
as.data.frame(table(assemComplete$subtype.present))[2,2],
as.data.frame(table(assemContig$subtype.present))[2,2],
as.data.frame(table(assemAR$subtype.present))[2,2],
as.data.frame(table(assemBR$subtype.present))[2,2],
as.data.frame(table(assemAC$subtype.present))[2,2],
as.data.frame(table(assemBC$subtype.present))[2,2]))

casSummary <- casSummary%>%mutate(Genomes.with.subtypes.and.arrays = c(as.data.frame(table(assemA$array.and_subtype))[4,2],
as.data.frame(table(assemB$array.and_subtype))[4,2],
as.data.frame(table(assemRep$array.and_subtype))[4,2],
as.data.frame(table(assemComplete$array.and_subtype))[4,2],
as.data.frame(table(assemContig$array.and_subtype))[4,2],
as.data.frame(table(assemAR$array.and_subtype))[4,2],
as.data.frame(table(assemBR$array.and_subtype))[4,2],
as.data.frame(table(assemAC$array.and_subtype))[4,2],
as.data.frame(table(assemBC$array.and_subtype))[4,2]))

casSummary <- casSummary%>%mutate(number.of.arrays = c(mean(na.omit(as.numeric(assemA$n.of.arrays))),
mean(na.omit(as.numeric(assemB$n.of.arrays))),
mean(na.omit(as.numeric(assemRep$n.of.arrays))),
mean(na.omit(as.numeric(assemComplete$n.of.arrays))),
mean(na.omit(as.numeric(assemContig$n.of.arrays))),
mean(na.omit(as.numeric(assemAR$n.of.arrays))),
mean(na.omit(as.numeric(assemBR$n.of.arrays))),
mean(na.omit(as.numeric(assemAC$n.of.arrays))),
mean(na.omit(as.numeric(assemBC$n.of.arrays)))))










tmp <- t(casSummary)
colnames(tmp) <- tmp["Category",]
tmp <- as.data.frame(tmp)
tmp <- tmp%>%filter(Archaea != "Archaea")%>%select(Archaea, Bacteria)%>%mutate(All = as.numeric(as.character(Archaea)) + as.numeric(as.character(Bacteria)))%>%select(All)
tmp <- t(tmp)
tmp <- as.data.frame(tmp)
tmp <- cbind(row.names(tmp), tmp)

colnames(tmp) <- colnames(casSummary)

casSummary <- rbind(casSummary, tmp)

casSummary <- casSummary%>%mutate(number.of.arrays = ifelse(Category == "All", NA, number.of.arrays))%>%mutate(number.of.arrays = ifelse(Category == "All", mean(na.omit(number.of.arrays)), number.of.arrays))

casSummary <- casSummary%>%
  mutate(p.arrays = round(Genomes.with.arrays/Genome.count*100, 2))%>%
  mutate(p.subtypes = round(Genomes.with.subtypes.identified/Genome.count*100, 2))%>%
  mutate(p.subtypes_and_arrays = round(Genomes.with.subtypes.and.arrays/Genome.count*100, 2))


subtypeSummary <- as.data.frame(table(assem$subtype.list))
subtypeSummary <- subtypeSummary%>%mutate(subtype.counts = str_count(Var1, "_") + 1)
singleSubSum <- subtypeSummary%>%filter(subtype.counts == 1)%>%filter(Var1 != "-")%>%mutate(subtype = Var1)%>%select(subtype, Freq)



subtypeSummaryRep <- as.data.frame(table(assemRep$subtype.list))
subtypeSummaryRep <- subtypeSummaryRep%>%mutate(subtype.counts = str_count(Var1, "_") + 1)
singleSubSumRep <- subtypeSummaryRep%>%filter(subtype.counts == 1)%>%filter(Var1 != "-")%>%mutate(subtype = Var1)%>%select(subtype, Freq)


sum(singleSubSum$Freq)

subtypeCounts <- as.data.frame(table(assem$subtype.count))


a <- casSummary%>%select(Category, Genome.count, p.arrays, p.subtypes, p.subtypes_and_arrays)
b <- casSummary%>%select(Category, Genome.count, number.of.arrays, Genomes.with.arrays)


casDat <-  read.table("~/Desktop/Project/identifycasproteins/cas_genes_refseq_83.txt", header = T, sep = '\t', comment.char = '', as.is = T, fill = T, quote = "")


casDat <- casDat%>%mutate(keep = ifelse(group == "new", ifelse(Gene_name == "csa3", "n", ifelse(Gene_name == "WYL", "n", ifelse(Gene_name == "DinG", "n", ifelse(Gene_name == "DEDDh", "n", ifelse(Gene_name == "PD-DExK", "n", ifelse(Gene_name == "PrimPol", "n", ifelse(Gene_name == "RT", "n", keep))))))), keep))%>%filter(keep == "y")


table(casDat$group)

nrow(casDat%>%filter(group == "Not found")%>%filter(!is.na(Gene_name)))

ggplot() +
  geom_bar(data = subset(casDat, group == "-"), aes(x = Gene_name, y = ..count..))  + 
  theme(axis.text.x=element_text(angle=90,hjust=1))


newCasDat <- casDat%>%filter(group == "new")

ggplot() +
  geom_bar(data = newCasDat, aes(x = Gene_name, y = ..count..))  + 
  theme(axis.text.x=element_text(angle=90,hjust=1))


newCasCount <- as.data.frame(table(newCasDat$Gene_name))

newCasCount <- newCasCount%>%filter(Freq > 500)%>%mutate(Var1 = ifelse(grepl("cmr5", Var1), "cmr5", ifelse(grepl("csm3", Var1), "csm3", ifelse(grepl("cas01", Var1), "cas1", as.character(Var1)))))%>%mutate(order.number = c(1,10, 2, 3, 4, 5, 6, 7,8, 9, 15, 14, 13, 12))%>%arrange(order.number)


newCasCount$Var1 <- factor(newCasCount$Var1, levels = newCasCount$Var1[order(newCasCount$order.number)])

ggplot() +
  geom_bar(data = newCasCount, aes(x = Var1, y = Freq), stat="identity")  + 
  theme(axis.text.x=element_text(angle=90,hjust=1)) +
  xlab("Cas Gene") +
  ylab("Frequency of Cas Genes") +
  theme_classic()

write.table(x = spacerSubtypes, file = "~/Desktop/Project/identifycasproteins/single_system_CRISPRDetect_refseq_83.fna", quote = F, sep = "\t", row.names = F, col.names = F)

```



```{r tmp}
##Import data
##import the hmmsearch output
hmm <- read.table("~/Desktop/Project/identifycasproteins/predicted_cas_genes_in_archaea_of_refseq83.txt", comment.char = "", sep = "", header = F, fill = T, col.names=c("target_name",        "accession",  "query_name",           "accession.2",    "E-value",  "score",  "bias",   "E-value.2",  "score.2",  "bias.2",   "exp", "reg", "clu" , "ov" ,"env" ,"dom" ,"rep", "inc" ,"description_of_target"))

##remove columns that are not hmm output (such as column names or comments)
hmm <- hmm%>%filter(description_of_target == "-")

rps <- read.table("~/Desktop/Project/identifycasproteins/archaea_refseq_83.cdd.out", comment.char = "#", sep = "", header = F, fill = T, col.names = c( "query_acc.ver", "subject_acc.ver", "identity_%", "alignment_length", "mismatches", "gap_opens", "q.start", "q.end", "s.start", "s.end", "evalue", "bit_score"))

##import the contig names and file locations
contig.lookup <- read.table("~/Desktop/Project/identifycasproteins/tmp.protein_name_lookup.txt", comment.char = "", fill = T)
colnames(contig.lookup) <- c("filepath", "target_name")
contig.lookup <- unique(contig.lookup)

##import models and subtypes table
genes <- read.table("~/Desktop/Project/identifycasproteins/cas_genes_and_systems.txt", header = T, sep = '\t', comment.char = '', as.is = T, fill = T)

##import signature cas gene lookup table
signatures <- read.table("~/Desktop/Project/identifycasproteins/models_and_systems.txt", header = T, sep = '\t', comment.char = '', as.is = T, fill = T)


setwd('~/Desktop/Project/identifycasproteins/')

##set the names for importing files and variables to use in this section. The files are not currently on the laptop.
args <- c('bacteria_refseq_79.sig_columns.all.tab',
          'bacteria_refseq_79.cdd.nr.tab',
          'bacteria_refseq_79.filelocations.txt',
          'cas_genes_and_systems.txt', 
          'models_and_systems.txt',
          'Bacteria')
cas <- read.table(args[1], header = T, comment.char = '#', as.is = T, fill = T)
cdd <- read.table(args[2], header = T, sep = '\t', comment.char = '', as.is = T, fill = T)
genomes <- read.table(args[3], header = T, sep = '\t', comment.char = '', fill = T)
genes <- read.table(args[4], header = T, sep = '\t', comment.char = '', as.is = T, fill = T)
signatures <- read.table(args[5], header = T, sep = '\t', comment.char = '', as.is = T, fill = T)
genes.dat <- data.frame(Accession = cas$Protein_ID, cas.id = cas$Cas_Model, 
                        gene.name = vector(mode='character', length=length(cas$Protein_ID)), putative.system = vector(mode='character', length=length(cas$Protein_ID)), 
                        cas.e.value.num= cas$Cas_E_value, cdd.e.value.num = vector(mode='numeric', length=length(cas$Protein_ID)),
                        genome = vector(mode='character', length=length(cas$Protein_ID)), keep.y.n = vector(mode='logical', length=length(cas$Protein_ID)), 
                        also.in.genomes = vector(mode='character', length=length(cas$Protein_ID)))

genomes <- genomes%>%filter(!is.na(Protein_ID))
genomes <- genomes%>%filter(Protein_ID != "")
#cas <- cas[,c(1,3,5,19,20)]
# x <- c()
#i <- 1
#for(i in 1:nrow(genomes)){
 # print(i)
 
#  if(is.na(genomes[i,2])){
#    x <- c(x, i)
 # }else{
 #   print(x)
 #   genomes[x,2] <- genomes[i,2]
#     x <- c()
#  }
#}
#write.table(genomes, file = "~/Desktop/Project/identifycasproteins/bacteria_refseq_79.filelocations.txt", sep = "\t", quote = F, row.names = F, col.names = T)

dat <- left_join(cas, cdd, by = "Protein_ID")
colnames(genes) <- c("Cas_Model", "Gene", "Subtype")
dat <- left_join(dat, genes, by = "Cas_Model")
#colnames(genomes) <- c("Protein_ID", "File_path")

colnames(dat) <- c( "Protein_ID", "Cas_Model", "Cas_E_value.num", "Protein_Description", "Species_Name", "GI", "Cdd_ID", "Cdd_E_value.num", "Gene", "Subtype")

dat <- transform.df(dat)
dat <- dat%>%mutate(cas_protein = ifelse(is.na(Cdd_E_value.num) , T, ifelse(Cdd_E_value.num*100 < Cas_E_value.num, ifelse(Cas_E_value.num > 1e-15, F, T), T)))%>%filter(cas_protein == T)%>%select(-cas_protein, -Cdd_E_value.num, -Cdd_ID, -GI)%>%filter(Gene != "DEDDh")%>%filter(Gene != "WYL")%>%filter(Gene != "PD-DExK")%>%filter(Gene != "RT")%>%filter(Gene != "PrimPol")

colnames(signatures) <- c("Cas_Model", "Cas_Gene_Name", "CRISPR_Cas_System", "Description", "Subtype.signature.genes")

dat <- left_join(dat, signatures, by = "Cas_Model")

dat <- dat%>%mutate(keep.gene = ifelse(Cas_Gene_Name == "DinG", ifelse(Cas_E_value.num < 1e-30, T, F), T))%>%filter(keep.gene == T)

table(dat$Cas_Gene_Name)

dat.reduced <- dat%>%filter(Cas_Gene_Name != "csx3")%>%filter(Cas_Gene_Name != "csx20")%>%filter(Cas_Gene_Name != "csx18")%>%filter(Cas_Gene_Name != "csx16")%>%filter(Cas_Gene_Name != "csx15")%>%filter(Cas_Gene_Name != "cas6")%>%filter(Cas_Model != "COG1468")%>%filter(Cas_Model != "cd09637")%>%filter(Cas_Gene_Name != "cas2")%>%filter(Cas_Gene_Name != "cas1")
dat.reduced.2 <- dat%>%filter(Subtype.signature.genes != "-")

tt <- dat.reduced.2

tt <- merge(x = tt, y = genomes, by = "Protein_ID", all.x = TRUE)

dat <- left_join(genomes, dat.reduced.2, by = "Protein_ID")



unique(dat$Gene)

table(dat$cas_protein)
##


##




x1 <- vector(mode = 'character', length = length(unique(genomes$V2)))
for(i in 1:length(unique(genomes$V2))){
  y <- strsplit(unique(genomes[i,2]), '/')[[1]][14]
  x1[i] <- y
}
x2 <- vector(mode = 'character', length = length(unique(genomes$V2)))
for(i in 1:length(unique(genomes$V2))){
  y <- paste('/mnt/SSD/CD_Tracr_Cas9',paste(strsplit(substr(genomes[i,2], 57, nchar(genomes[i,2])),'/')[[1]][1:4], collapse = '/'),'/*repeat.txt', sep = '')
  x2[i] <- y
}
x3 <- vector(mode = 'character', length = length(unique(genomes$V2)))
for(i in 1:length(unique(genomes$V2))){
  y <- substr(unique(genomes[i,2]),12,nchar(genomes[i,2]))
  x3[i] <- y
}
genomes.dat <- data.frame(genome =x1,
                          genes = vector(mode = 'character', length=length(unique(genomes$V2))),
                          basic.genes.present.y.n = vector(mode = 'logical', length=length(unique(genomes$V2))), single.system.y.n =vector(mode = 'logical', length=length(unique(genomes$V2))),
                          subtypes = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          fastafile_path =  x3,
                          CRISPRDetect_path = x2,
                          Classification = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          cas.e.values = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          cdd.e.values = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          cas.models = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          all.proteins = genomes$V1)
genes[genes[,3]=='CAS-III-C' & genes[,2]=='cas10', 2] <- 'cas10c'
gene.summary.dat <- data.frame(gene = unique(genes$Gene), systems = vector(mode = 'character', length = length(unique(genes$Gene))), unique.y.n = vector(mode = 'logical', length = length(unique(genes$Gene))),subtypes.y.n = vector(mode = 'logical', length = length(unique(genes$Gene))), models = vector(mode = 'character', length = length(unique(genes$Gene))) )
gene.summary.dat <- transform.df(gene.summary.dat)
#unique(genes$Gene)[substr(unique(genes$Gene), 1, 5)=='cas10']
for(i in 1:length(unique(genes$Gene))){
  cas.gene <- unique(genes$Gene)[i]
  dat <- genes[genes[,2]==cas.gene,]
  gene.summary.dat[i, 2] <- paste(sort(unique(dat[,3])), collapse = ',')
  gene.summary.dat[i, 5] <- paste(sort(unique(dat[,1])), collapse = ',')
  if(length(unique(dat[,3]))==1){
    gene.summary.dat[i, 3] <- T
  }else{
    gene.summary.dat[i, 3] <- F
  }
  
}

gene.summary.dat <- gene.summary.dat[gene.summary.dat[,2]!='CAS-I,CAS-III',]
gene.summary.dat$subtypes.y.n <- T
gene.summary.dat[gene.summary.dat[,2]=='CAS-I', 4] <- F
gene.summary.dat[gene.summary.dat[,2]=='CAS-II', 4] <- F
gene.summary.dat[gene.summary.dat[,2]=='CAS-III', 4] <- F
gene.summary.dat[gene.summary.dat[,1]=='csa3', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='DinG', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf2gr7', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf3gr5', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf4gr11', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf5gr6', 3] <- F

#------------------------------------commands to run single core---------------------------------------------#


genes.dat <- transform.df(genes.dat)
genomes.dat <- transform.df(genomes.dat)
gene.summary.dat <- transform.df(gene.summary.dat)

ptm <- proc.time()
for(i in 1:length(cas[,1])){
  genes.dat[i,] <- write.genes.dat(i, cas, cdd, genes.dat, signatures, gene.summary.dat, genomes, print.i = T, genomes.dat)
}

proc.time() - ptm
genes.dat[genes.dat[,3]=='DinG',3] <- ''
genes.dat[genes.dat[,3]=='casR',3] <- ''
genes.dat[genes.dat[,3]=='cas3HD',3] <- ''
genes.dat[genes.dat[,3]=='cmr4gr7',3] <- ''
genes.dat[genes.dat[,3]=='cmr6gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm3gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm6gr7',3] <- ''
genes.dat[genes.dat[,3]=='csx1gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm5gr7',3] <- ''
genes.dat[genes.dat[,3]=='cmr1gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm4gr5',3] <- ''
#genes.dat[genes.dat] <- ''
genes.dat[genes.dat[,3]=='csm6',3] <- ''
genes.dat[genes.dat[,3]=='csx1',3] <- ''
genes.dat[genes.dat[,3]=='DEDDh',3] <- ''

y <- grep('cas7', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas7'
  }
  
}
y <- grep('cas8b', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas8b'
  }  
}
y <- grep('cas5', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas5'
  }  
}
y <- grep('cas3', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas3'
  }
}

a <- c()
for(i in 1:length(genomes.dat[,1])){
  genome <- genomes.dat[i,1]
  x <- grep(genome, genes.dat$genome)

  dat <- genes.dat[x,]
  
  uniq.uids <- unique(dat[,1])
  uniq.dat <- dat[1:length(uniq.uids),]
  uniq.dat <- transform.df(uniq.dat)
  for(j in 1:length(uniq.uids)){
    x <- uniq.uids[j]
    tmp.dat <- dat[dat[,1]==x,]
    tmp.dat <- tmp.dat[order(tmp.dat[,5]),]
    if(length(unique(tmp.dat[,3]))<2){
      uniq.dat[j,] <- tmp.dat[1,]
    }else{
      uniq.dat[j,] <- tmp.dat[1,]
      a <- c(a, paste(unique(tmp.dat[,3]), collapse = ','))
    }
  }
  genomes.dat[i,8] <- args[6]
  gene.list <- uniq.dat[,3]
  genomes.dat[i,2] <- paste(gene.list, collapse = ',')
  genomes.dat[i,5] <- paste(sort(unique(uniq.dat[,4])), collapse = ',')
  genomes.dat[i,9] <- paste(uniq.dat[,5], collapse = ',')
  genomes.dat[i,10] <- paste(sort(unique(uniq.dat[,5])), collapse = ',')
  genomes.dat[i,11] <- paste(uniq.dat[,2], collapse = ',')
  genomes.dat[i,12] <- paste(uniq.dat[,1], collapse = ',')
  x <-match('cas1', uniq.dat[,3])
  y <-match('cas2', uniq.dat[,3])
  
  if(!is.na(x) & !is.na(y)){
    genomes.dat[i,3] <- T
  }
  if(genomes.dat[i,5]==''){
    x <- grep('cas9', uniq.dat$gene.name)
    y <- grep('cas4', uniq.dat$gene.name)
    z <- grep('csn2', uniq.dat$gene.name)
    if(length(x)>0){
      if(length(y)>0){
        genomes.dat[i,5] <- ',CAS-II-B'
      }else if(length(z)==0){
        genomes.dat[i,5] <- ',CAS-II-C'
        
      }
    }
  }
  if(genomes.dat[i,5]==',CAS-I-E'){
    y <- grep('cas4', uniq.dat$gene.name)
    if(length(x)>0){
      genomes.dat[i,5] <- ',CAS-I-E, OTHER?'
      genomes.dat[i,4] <- F
    }
  }
  if(genomes.dat[i,5]==',CAS-I-F'){
    y <- grep('cas4', uniq.dat$gene.name)
    if(length(x)>0){
      genomes.dat[i,5] <- ',CAS-I-F, OTHER?'
      genomes.dat[i,4] <- F
    }
  }
  x <- strsplit(genomes.dat[i, 5], ',')
  if(length(x[[1]])!=0){
    if(x[[1]][1]==''){
      x[[1]] <- x[[1]][2:length(x[[1]])]
    }
    if(length(x[[1]])==1){
      genomes.dat[i,4] <- T
      
    }else{
      genomes.dat[i,4] <- F
    }
  }else{
    genomes.dat[i,4] <- F
    
  }
  if(genomes.dat[i,2]==F){
    #genomes.dat[i,7] <- ''
    
  }
  
  
}

a.uniq <- data.frame(list = unique(a), counts.num = vector(mode = 'numeric', length = length( unique(a))))
a.uniq <- transform.df(a.uniq)
for(i in 1:length(a.uniq[,1])){
  x <- a.uniq[i,1]
  y <- a[a==x]
  a.uniq[i,2] <- length(y)
}
write.table(a.uniq, 'uids_matched_with_multiple_cas_models_archaea_refseq_79.txt', quote = F, row.names = F)
write.table(genes.dat, 'cas_protein_data_archaea_refseq_79.txt', quote = F, row.names = F, sep = '\t')
write.table(genomes.dat, 'genomes_with_cas_proteins_archaea_refseq_79.txt', quote = F, row.names = F, sep = '\t')
write.table(genomes.dat[genomes.dat[,4]==T, ], 'genomes_with_single_system_archaea_refseq_79.txt', quote = F, row.names = F, sep = '\t')


```







```{r identify_cas_genomes, eval=F, echo=T}
##Import data
setwd('~/Desktop/Project/identifycasproteins/')

##set the names for importing files and variables to use in this section. The files are not currently on the laptop.
args <- c('bacteria_refseq_79.sig_columns.all.tab',
          'bacteria_refseq_79.cdd.nr.tab',
          'bacteria_refseq_79.filelocations.txt',
          'cas_genes_and_systems.txt', 
          'models_and_systems.txt',
          'Bacteria')
cas <- read.table(args[1], header = T, comment.char = '#', as.is = T, fill = T)
cdd <- read.table(args[2], header = T, sep = '\t', comment.char = '', as.is = T, fill = T)
genomes <- read.table(args[3], header = T, sep = '\t', comment.char = '', fill = T)
genes <- read.table(args[4], header = T, sep = '\t', comment.char = '', as.is = T, fill = T)
signatures <- read.table(args[5], header = T, sep = '\t', comment.char = '', as.is = T, fill = T)
genes.dat <- data.frame(Accession = cas$Protein_ID, cas.id = cas$Cas_Model, 
                        gene.name = vector(mode='character', length=length(cas$Protein_ID)), putative.system = vector(mode='character', length=length(cas$Protein_ID)), 
                        cas.e.value.num= cas$Cas_E_value, cdd.e.value.num = vector(mode='numeric', length=length(cas$Protein_ID)),
                        genome = vector(mode='character', length=length(cas$Protein_ID)), keep.y.n = vector(mode='logical', length=length(cas$Protein_ID)), 
                        also.in.genomes = vector(mode='character', length=length(cas$Protein_ID)))

genomes <- genomes%>%filter(!is.na(Protein_ID))
genomes <- genomes%>%filter(Protein_ID != "")
#cas <- cas[,c(1,3,5,19,20)]
# x <- c()
#i <- 1
#for(i in 1:nrow(genomes)){
 # print(i)
 
#  if(is.na(genomes[i,2])){
#    x <- c(x, i)
 # }else{
 #   print(x)
 #   genomes[x,2] <- genomes[i,2]
#     x <- c()
#  }
#}
#write.table(genomes, file = "~/Desktop/Project/identifycasproteins/bacteria_refseq_79.filelocations.txt", sep = "\t", quote = F, row.names = F, col.names = T)

dat <- left_join(cas, cdd, by = "Protein_ID")
colnames(genes) <- c("Cas_Model", "Gene", "Subtype")
dat <- left_join(dat, genes, by = "Cas_Model")
#colnames(genomes) <- c("Protein_ID", "File_path")

colnames(dat) <- c( "Protein_ID", "Cas_Model", "Cas_E_value.num", "Protein_Description", "Species_Name", "GI", "Cdd_ID", "Cdd_E_value.num", "Gene", "Subtype")

dat <- transform.df(dat)
dat <- dat%>%mutate(cas_protein = ifelse(is.na(Cdd_E_value.num) , T, ifelse(Cdd_E_value.num*100 < Cas_E_value.num, ifelse(Cas_E_value.num > 1e-15, F, T), T)))%>%filter(cas_protein == T)%>%select(-cas_protein, -Cdd_E_value.num, -Cdd_ID, -GI)%>%filter(Gene != "DEDDh")%>%filter(Gene != "WYL")%>%filter(Gene != "PD-DExK")%>%filter(Gene != "RT")%>%filter(Gene != "PrimPol")

colnames(signatures) <- c("Cas_Model", "Cas_Gene_Name", "CRISPR_Cas_System", "Description", "Subtype.signature.genes")

dat <- left_join(dat, signatures, by = "Cas_Model")

dat <- dat%>%mutate(keep.gene = ifelse(Cas_Gene_Name == "DinG", ifelse(Cas_E_value.num < 1e-30, T, F), T))%>%filter(keep.gene == T)

table(dat$Cas_Gene_Name)

dat.reduced <- dat%>%filter(Cas_Gene_Name != "csx3")%>%filter(Cas_Gene_Name != "csx20")%>%filter(Cas_Gene_Name != "csx18")%>%filter(Cas_Gene_Name != "csx16")%>%filter(Cas_Gene_Name != "csx15")%>%filter(Cas_Gene_Name != "cas6")%>%filter(Cas_Model != "COG1468")%>%filter(Cas_Model != "cd09637")%>%filter(Cas_Gene_Name != "cas2")%>%filter(Cas_Gene_Name != "cas1")
dat.reduced.2 <- dat%>%filter(Subtype.signature.genes != "-")

tt <- dat.reduced.2

tt <- merge(x = tt, y = genomes, by = "Protein_ID", all.x = TRUE)

dat <- left_join(genomes, dat.reduced.2, by = "Protein_ID")



unique(dat$Gene)

table(dat$cas_protein)
##


##




x1 <- vector(mode = 'character', length = length(unique(genomes$V2)))
for(i in 1:length(unique(genomes$V2))){
  y <- strsplit(unique(genomes[i,2]), '/')[[1]][14]
  x1[i] <- y
}
x2 <- vector(mode = 'character', length = length(unique(genomes$V2)))
for(i in 1:length(unique(genomes$V2))){
  y <- paste('/mnt/SSD/CD_Tracr_Cas9',paste(strsplit(substr(genomes[i,2], 57, nchar(genomes[i,2])),'/')[[1]][1:4], collapse = '/'),'/*repeat.txt', sep = '')
  x2[i] <- y
}
x3 <- vector(mode = 'character', length = length(unique(genomes$V2)))
for(i in 1:length(unique(genomes$V2))){
  y <- substr(unique(genomes[i,2]),12,nchar(genomes[i,2]))
  x3[i] <- y
}
genomes.dat <- data.frame(genome =x1,
                          genes = vector(mode = 'character', length=length(unique(genomes$V2))),
                          basic.genes.present.y.n = vector(mode = 'logical', length=length(unique(genomes$V2))), single.system.y.n =vector(mode = 'logical', length=length(unique(genomes$V2))),
                          subtypes = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          fastafile_path =  x3,
                          CRISPRDetect_path = x2,
                          Classification = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          cas.e.values = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          cdd.e.values = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          cas.models = vector(mode = 'character', length=length(unique(genomes$V2))), 
                          all.proteins = genomes$V1)
genes[genes[,3]=='CAS-III-C' & genes[,2]=='cas10', 2] <- 'cas10c'
gene.summary.dat <- data.frame(gene = unique(genes$Gene), systems = vector(mode = 'character', length = length(unique(genes$Gene))), unique.y.n = vector(mode = 'logical', length = length(unique(genes$Gene))),subtypes.y.n = vector(mode = 'logical', length = length(unique(genes$Gene))), models = vector(mode = 'character', length = length(unique(genes$Gene))) )
gene.summary.dat <- transform.df(gene.summary.dat)
#unique(genes$Gene)[substr(unique(genes$Gene), 1, 5)=='cas10']
for(i in 1:length(unique(genes$Gene))){
  cas.gene <- unique(genes$Gene)[i]
  dat <- genes[genes[,2]==cas.gene,]
  gene.summary.dat[i, 2] <- paste(sort(unique(dat[,3])), collapse = ',')
  gene.summary.dat[i, 5] <- paste(sort(unique(dat[,1])), collapse = ',')
  if(length(unique(dat[,3]))==1){
    gene.summary.dat[i, 3] <- T
  }else{
    gene.summary.dat[i, 3] <- F
  }
  
}

gene.summary.dat <- gene.summary.dat[gene.summary.dat[,2]!='CAS-I,CAS-III',]
gene.summary.dat$subtypes.y.n <- T
gene.summary.dat[gene.summary.dat[,2]=='CAS-I', 4] <- F
gene.summary.dat[gene.summary.dat[,2]=='CAS-II', 4] <- F
gene.summary.dat[gene.summary.dat[,2]=='CAS-III', 4] <- F
gene.summary.dat[gene.summary.dat[,1]=='csa3', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='DinG', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf2gr7', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf3gr5', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf4gr11', 3] <- F
gene.summary.dat[gene.summary.dat[,1]=='csf5gr6', 3] <- F

#------------------------------------commands to run single core---------------------------------------------#


genes.dat <- transform.df(genes.dat)
genomes.dat <- transform.df(genomes.dat)
gene.summary.dat <- transform.df(gene.summary.dat)

ptm <- proc.time()
for(i in 1:length(cas[,1])){
  genes.dat[i,] <- write.genes.dat(i, cas, cdd, genes.dat, signatures, gene.summary.dat, genomes, print.i = T, genomes.dat)
}

proc.time() - ptm
genes.dat[genes.dat[,3]=='DinG',3] <- ''
genes.dat[genes.dat[,3]=='casR',3] <- ''
genes.dat[genes.dat[,3]=='cas3HD',3] <- ''
genes.dat[genes.dat[,3]=='cmr4gr7',3] <- ''
genes.dat[genes.dat[,3]=='cmr6gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm3gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm6gr7',3] <- ''
genes.dat[genes.dat[,3]=='csx1gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm5gr7',3] <- ''
genes.dat[genes.dat[,3]=='cmr1gr7',3] <- ''
genes.dat[genes.dat[,3]=='csm4gr5',3] <- ''
#genes.dat[genes.dat] <- ''
genes.dat[genes.dat[,3]=='csm6',3] <- ''
genes.dat[genes.dat[,3]=='csx1',3] <- ''
genes.dat[genes.dat[,3]=='DEDDh',3] <- ''

y <- grep('cas7', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas7'
  }
  
}
y <- grep('cas8b', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas8b'
  }  
}
y <- grep('cas5', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas5'
  }  
}
y <- grep('cas3', genes.dat[,3])
if(length(y)!=0){
  for(i in 1:length(y)){
    genes.dat[y[i],3] <- 'cas3'
  }
}

a <- c()
for(i in 1:length(genomes.dat[,1])){
  genome <- genomes.dat[i,1]
  x <- grep(genome, genes.dat$genome)

  dat <- genes.dat[x,]
  
  uniq.uids <- unique(dat[,1])
  uniq.dat <- dat[1:length(uniq.uids),]
  uniq.dat <- transform.df(uniq.dat)
  for(j in 1:length(uniq.uids)){
    x <- uniq.uids[j]
    tmp.dat <- dat[dat[,1]==x,]
    tmp.dat <- tmp.dat[order(tmp.dat[,5]),]
    if(length(unique(tmp.dat[,3]))<2){
      uniq.dat[j,] <- tmp.dat[1,]
    }else{
      uniq.dat[j,] <- tmp.dat[1,]
      a <- c(a, paste(unique(tmp.dat[,3]), collapse = ','))
    }
  }
  genomes.dat[i,8] <- args[6]
  gene.list <- uniq.dat[,3]
  genomes.dat[i,2] <- paste(gene.list, collapse = ',')
  genomes.dat[i,5] <- paste(sort(unique(uniq.dat[,4])), collapse = ',')
  genomes.dat[i,9] <- paste(uniq.dat[,5], collapse = ',')
  genomes.dat[i,10] <- paste(sort(unique(uniq.dat[,5])), collapse = ',')
  genomes.dat[i,11] <- paste(uniq.dat[,2], collapse = ',')
  genomes.dat[i,12] <- paste(uniq.dat[,1], collapse = ',')
  x <-match('cas1', uniq.dat[,3])
  y <-match('cas2', uniq.dat[,3])
  
  if(!is.na(x) & !is.na(y)){
    genomes.dat[i,3] <- T
  }
  if(genomes.dat[i,5]==''){
    x <- grep('cas9', uniq.dat$gene.name)
    y <- grep('cas4', uniq.dat$gene.name)
    z <- grep('csn2', uniq.dat$gene.name)
    if(length(x)>0){
      if(length(y)>0){
        genomes.dat[i,5] <- ',CAS-II-B'
      }else if(length(z)==0){
        genomes.dat[i,5] <- ',CAS-II-C'
        
      }
    }
  }
  if(genomes.dat[i,5]==',CAS-I-E'){
    y <- grep('cas4', uniq.dat$gene.name)
    if(length(x)>0){
      genomes.dat[i,5] <- ',CAS-I-E, OTHER?'
      genomes.dat[i,4] <- F
    }
  }
  if(genomes.dat[i,5]==',CAS-I-F'){
    y <- grep('cas4', uniq.dat$gene.name)
    if(length(x)>0){
      genomes.dat[i,5] <- ',CAS-I-F, OTHER?'
      genomes.dat[i,4] <- F
    }
  }
  x <- strsplit(genomes.dat[i, 5], ',')
  if(length(x[[1]])!=0){
    if(x[[1]][1]==''){
      x[[1]] <- x[[1]][2:length(x[[1]])]
    }
    if(length(x[[1]])==1){
      genomes.dat[i,4] <- T
      
    }else{
      genomes.dat[i,4] <- F
    }
  }else{
    genomes.dat[i,4] <- F
    
  }
  if(genomes.dat[i,2]==F){
    #genomes.dat[i,7] <- ''
    
  }
  
  
}

a.uniq <- data.frame(list = unique(a), counts.num = vector(mode = 'numeric', length = length( unique(a))))
a.uniq <- transform.df(a.uniq)
for(i in 1:length(a.uniq[,1])){
  x <- a.uniq[i,1]
  y <- a[a==x]
  a.uniq[i,2] <- length(y)
}
write.table(a.uniq, 'uids_matched_with_multiple_cas_models_archaea_refseq_79.txt', quote = F, row.names = F)
write.table(genes.dat, 'cas_protein_data_archaea_refseq_79.txt', quote = F, row.names = F, sep = '\t')
write.table(genomes.dat, 'genomes_with_cas_proteins_archaea_refseq_79.txt', quote = F, row.names = F, sep = '\t')
write.table(genomes.dat[genomes.dat[,4]==T, ], 'genomes_with_single_system_archaea_refseq_79.txt', quote = F, row.names = F, sep = '\t')


```


```{r crispr_models_table, results='asis'}
crispr_models_data <- read.table("~/Desktop/Project/identifycasproteins/genes_and_systems_summary.txt", as.is = T, sep = '\t', comment.char = "", header = T)
crispr_models_data <- crispr_models_data[crispr_models_data[,3]==T,]
tab_7 <- xtable(crispr_models_data[,1:4], caption=('Cas genes assignment to CRISPR systems'))
print(tab_7, type="latex", caption.placement='top', comment=FALSE)

```

Data at this point has been assigned CRISPR systems and matched up with CRISPRDetect files. The script from here is working on producing some summary results. 

I will need to take the genomes that look like they have CRISPR systems and are missing CRISPRDetect files and rerun these to see what is going on. I many cases the support for a CRISPR systems is strong based on the genes found.



### This is currently broken and will need fixing/doing again better ###
```{r setup_cont, cache=T, cache.lazy=T, results='asis', eval=F}
#-----------------------------------Setup and import data-----------------------------------#
##import genome data
genomes <- read.table("~/Desktop/Project/identifycasproteins/Refseq_79/refseq_79.genomes_summary.txt", header = T, sep = '\t', as.is = T, comment.char = '')

##count the number of putative cas genes in each genome
gene.count <- vector(mode = 'numeric', length = length(genomes[,1]))
for(i in 1:length(genomes[,1])){
  x <- strsplit(genomes[i,2], ',')
  y <- length(x[[1]])
  gene.count[i] <- y
}

##Identify I-E and I-F genomes that were excluded due to the presence of DinG in genome and include in list of genomes with a single subtype.
for(i in 1:nrow(genomes)){
  if(genomes[i, 5]==",CAS-I-E, OTHER?"){
    genomes[i, 5] <- ",CAS-I-E"
  }else if(genomes[i, 5]==",CAS-I-F, OTHER?"){
    genomes[i, 5] <- ",CAS-I-F"
  }else if(genomes[i, 5]=="CAS-I-E, OTHER?"){
    genomes[i, 5] <- "CAS-I-E"
  }else if(genomes[i, 5]=="CAS-I-F, OTHER?"){
    genomes[i, 5] <- "CAS-I-F"
  }
}


##count the number of subtypes in each genome
systems.count <- vector(mode = 'numeric', length = length(genomes[,1]))
for(i in 1:length(genomes[,1])){
  ##check a subtype is present
  if(!is.na(genomes[i,5])){
    ##account for blank space in some of the genome subtype lists
  if(substr(genomes[i,5],1,1)==','){
    x <- strsplit(genomes[i,5], ',')
    y <- length(x[[1]])
    systems.count[i] <- y-1
  }else{
    x <- strsplit(genomes[i,5], ',')
    y <- length(x[[1]])
    systems.count[i] <- y
  }
}
}



##remove genomes with no subtypes
systems.count.not.zero <- systems.count[systems.count!=0]


##count the number of genomes with x subtypes found (output in y)
y <- c(length(genomes[genomes[,2]==',',1]),length(systems.count.not.zero[systems.count.not.zero==1]),length(systems.count.not.zero[systems.count.not.zero==2]),length(systems.count.not.zero[systems.count.not.zero==3]),length(systems.count.not.zero[systems.count.not.zero==4]),length(systems.count.not.zero[systems.count.not.zero==5]))
x <- c('0', '1','2','3','4','5')

##generate list of subtypes to analyse
subtype.list <- c('CAS-I-A','CAS-I-B','CAS-I-C','CAS-I-D','CAS-I-E','CAS-I-F','CAS-I-U','CAS-II-A','CAS-II-B','CAS-II-C','CAS-III-A','CAS-III-B','CAS-III-C','CAS-III-D','CAS-IV','CAS-V')

##call function "subtypes.counts.funct" to count the number of times each subtype occurs
subtypes.dat <- subtypes.counts.funct(subtype.list, genomes )#subtype.list can be replaced with any other combination of subtypes in the form 'CAS-X-X,CAS-Y-Y,CAS-Z-Z,...'
subtypes.dat.arc <- subtypes.counts.funct(subtype.list, genomes[genomes[,8]=='Archaea',] )#subtype.list can be replaced with any other combination of subtypes in the form 'CAS-X-X,CAS-Y-Y,CAS-Z-Z,...'
subtypes.dat.bac <- subtypes.counts.funct(subtype.list, genomes[genomes[,8]=='Bacteria',] )#subtype.list can be replaced with any other combination of subtypes in the form 'CAS-X-X,CAS-Y-Y,CAS-Z-Z,...'
#hist(z, xlim = c(0, max(z)), breaks = max(z/5), main = 'Distribution of number of spacers in arrays', xlab = 'Number of spacers in an array', col = 'black')
arrays.dat <- c(length(genomes[genomes[,14]==1,1]),length(genomes[genomes[,14]==2,1]),length(genomes[genomes[,14]==3,1]),length(genomes[genomes[,14]==4,1]),length(genomes[genomes[,14]==5,1]),length(genomes[genomes[,14]==6,1]),length(genomes[genomes[,14]==7,1]),length(genomes[genomes[,14]==8,1]),length(genomes[genomes[,14]==9,1]))
arrays.dat.names <- c('1','2','3','4','5','6','7','8','9')

spacer.count <- c()
av.spacers.num <- c()
for(i in 1:length(genomes[,1])){
  a <- strsplit(genomes[i,13], ',')
  spacer.count <- c(spacer.count,as.numeric(a[[1]]))
  av.spacers.num <- c(av.spacers.num, mean(as.numeric(a[[1]])))
}
gene.count.num <- gene.count
genomes$array.num <- genomes$array.num/2
genomes <- cbind(genomes, av.spacers.num, gene.count.num)
genomes <- transform.df(genomes)
arrays.model <- lm(genomes$av.spacers.num~genomes$array.num )
summary.arrays.model <- summary(arrays.model)
ordered.gene.count <- gene.count[order(-gene.count)]
ordered.gene.count[1:10]
gene.frequency <- data.frame(table(ordered.gene.count))
ordered.gene.freq <- gene.frequency[order(-gene.frequency[,2]),]
tab_1 <- xtable(ordered.gene.freq[1:20,], caption = "Gene Frequency")
print(tab_1, type="latex", caption.placement='top', comment=FALSE)

system.freq <- data.frame(table(systems.count))
system.freq <- system.freq[system.freq[,1]!="0",]


barplot(system.freq[,2], names.arg = system.freq[,1], col = "Black", xlab = " number of systems", ylab = "Number of genomes", main = "Number of systems per genome")


tab_2 <- xtable(system.freq, caption = "Number of systems per genome")
print(tab_2, type="latex", caption.placement='top', comment=FALSE)
tab_3 <- xtable(subtypes.dat.arc, caption = "Number of archaeal genomes with each subtype")
print(tab_3, type="latex", caption.placement='top', comment=FALSE)
barplot(subtypes.dat.arc$counts, names.arg = subtypes.dat.arc$subtype, cex.names = 1, main = 'Archaeal Subtype Frequencies', col = "Black", las =2)
tab_4 <- xtable(subtypes.dat.bac, caption = "Number of bacterial genomes with each subtype")
print(tab_4, type="latex", caption.placement='top', comment=FALSE)
barplot(subtypes.dat.bac$counts, names.arg = subtypes.dat.bac$subtype, cex.names = 1, main = 'Bacterial Subtype Frequencies', col = "Black", las =2)
tab_5 <- xtable(subtypes.dat, caption = "Number of genomes with each subtype")
print(tab_5, type="latex", caption.placement='top', comment=FALSE)
barplot(subtypes.dat$counts, names.arg = subtypes.dat$subtype, cex.names = 1, main = 'Subtype Frequencies', col = "Black", las =2)
```


```{r import_summary_data, results='asis'}
summary.dat <- read.table("~/Desktop/Project/identifycasproteins/refseq_79.genomes_summary.txt", sep = '\t', fill = T, comment.char = '', header = T)

tt <- summary.dat%>%filter(single.system.y.n == T)

summary.dat <- transform.df(summary.dat)
summary.dat[,5] <- substr(summary.dat[,5], 2, nchar(summary.dat[,5]))
for(i in 1:nrow(summary.dat)){
  summary.dat[i,5] <- ifelse(substr(summary.dat[i,5], 1, 1)=="A", paste("C", summary.dat[i,5], sep = ""), summary.dat[i,5])
}

subtype.dat <- as.data.frame(table(as.factor(summary.dat$subtypes)))
colnames(subtype.dat) <- c("Subtypes", "Freq.num")
subtype.dat <- subtype.dat[subtype.dat[,1]!='',]
subtype.dat <- subtype.dat[subtype.dat[,1]!='ubtypes',]
subtype.dat <- transform.df(subtype.dat)
subtype.dat.common <- subtype.dat[subtype.dat[,2]>50,]
ordered.subtype.dat <- subtype.dat[order(-subtype.dat[,2]),]
ordered.subtype.dat <- transform.df(ordered.subtype.dat)
ordered.subtype.dat[1,1] <- "CAS-I-E"
ordered.subtype.dat[4,1] <- "CAS-I-F"
tab1 <- xtable(ordered.subtype.dat[1:50,], caption=('Frequency of different combinations of subtypes'))
print(tab1, type="latex", caption.placement='top', comment=FALSE)
subtype.dat <- subtype.dat[order(-subtype.dat$Freq),]
par(mar = c(10,4,4,2))


barplot(ordered.subtype.dat[1:20,2], names.arg = ordered.subtype.dat[1:20,1], col = "Black", las=2, ylab = "Frequency", main = "Number of genomes with different combinations of subtypes")

##Summary of the overall abundance and other data like that

xx <- data.frame(Archaea = c(453, 415, 313, 159, 220), Bacteria = c(31953, 18621, 6651, 3340, 3489))
rownames(xx) <- c("Number of genomes", "Number of genomes with cas", "Number of genomes with cas proteins and CRISPR arrays", "Number of genomes with a single system", "Number of genomes wth cas1 and cas2")
tab_6 <- xtable(xx, caption=('Genome analysis summary'))
print(tab_6, type="latex", caption.placement='top', comment=FALSE)

```


is run on the server
```{r append_gene_description_1, eval =F}
transform.df <- function(dat){
  for(i in 1:length(dat[1,])){
    if(substr(colnames(dat)[i],(nchar(colnames(dat)[i])-3),nchar(colnames(dat)[i]) )=='.num'){
      dat[,i] <- as.character(dat[,i])
      dat[,i] <- as.numeric(dat[,i])
    }else if(substr(colnames(dat)[i], nchar(colnames(dat)[i])-3, nchar(colnames(dat)[i]))=='.y.n'){
      dat[,i] <- as.logical(dat[,i])
    }else{
      dat[,i] <- as.character(dat[,i])
    }
  }
  return(dat)
}
library(dplyr)

print("genes.dat")
genes.dat <- read.table("/home/tomn/masters/Working/cas_proteins_refseq_79_gff_metadata.all.txt", header = T, sep = "\t", comment.char = "", fill = T, row.names = NULL)

print(head(genes.dat))

print("descriptions.dat")

descriptions.dat <- read.table("/home/tomn/masters/Working/gene_description.lookup", header = F, sep = "\t", comment.char = "", fill = T)
i.f.dat <- genes.dat#[genes.dat[,9]=="CAS-I-F",]
length(unique(i.f.dat$Accession))
i.f.dat <- i.f.dat[i.f.dat[,13]==T,]


colnames(descriptions.dat) <- c("Accession","Description")
descriptions.dat <- unique(descriptions.dat[,c(1,2)])

print("left_join")
tt <- left_join(i.f.dat, descriptions.dat, by = "Accession")

print("write.table")
write.table(tt, file = "/home/tomn/masters/Working/cas_proteins_refseq_79_gff_metadata.all.gene_descriptions.txt", sep = "\t", quote = F, row.names = F, col.names = T )


```

not running as the file is already written.
```{r cas_genes_summary, eval = F, include=T}
##import archaeal data
archaeal.genes <- read.table("~/Desktop/Project/identifycasproteins/cas_proteins_refseq_79_gff_metadata.archaea.txt", header = T, sep = "\t", fill = T, as.is = T)
Host <- rep("Archaea", nrow(archaeal.genes))

##append host info
archaeal.genes <- cbind(archaeal.genes, Host)

##import bacteria data
bacteria.genes <- read.table("~/Desktop/Project/identifycasproteins/cas_proteins_refseq_79_gff_metadata.bacteria_small.txt", header = T, sep = "\t", fill = T, as.is = T)
Host <- rep("Bacteria", nrow(bacteria.genes))
also.in.genomes <- rep(NA, nrow(bacteria.genes))
##aapend host info and the also.in.genomes
bacteria.genes <- cbind(bacteria.genes, also.in.genomes, Host)

##combine data
genes.dat <- rbind(archaeal.genes, bacteria.genes)

##write data to file

write.table(genes.dat, "~/Desktop/Project/identifycasproteins/cas_proteins_refseq_79_gff_metadata.all.txt", quote = F, row.names = F, col.names = T, sep = "\t")
```


Taking the results from the HMM search done by David using the archaeal cas9 hmms and checking the proteins that were found against the list of proteins I have identified as cas proteins.
```{r check_archaeal_cas9, eval = F, include=T}
##import list of cas proteins
archaeal.dat <- read.table("~/Desktop/Project/identifycasproteins/genomes_with_cas_proteins_archaea_refseq_79_gff_metadata.txt", header = T, sep = "\t", as.is = T, comment.char = "", fill = T)
##import davids hmmsearch results
cas9.dat <- read.csv("~/Downloads/archaeal_cas9.csv")
unique.match.y.n <- vector(mode = "logical", length = nrow(cas9.dat))
genome.found.y.n <- vector(mode = "logical", length = nrow(cas9.dat))
basic.genes.y.n <- vector(mode = "logical", length = nrow(cas9.dat))
any.cas.genes.y.n <- vector(mode = "logical", length = nrow(cas9.dat))
cas.genes <- vector(mode = "character", length = nrow(cas9.dat))
cas.gene.starts <- vector(mode = "character", length = nrow(cas9.dat))
cas.gene.stops <- vector(mode = "character", length = nrow(cas9.dat))
num.of.cas.genes.num <- vector(mode = "numeric", length = nrow(cas9.dat))
cas9.dat <- cbind(cas9.dat, unique.match.y.n, genome.found.y.n, basic.genes.y.n, any.cas.genes.y.n, cas.genes, cas.gene.starts, cas.gene.stops, num.of.cas.genes.num)
colnames(cas9.dat) <- c("PROTEIN_ID", "FS_EVAL.num", "ASSEMBLY",   "ORGANISM",   "CONTIG",     "HMM_NAME", "FS_SCORE.num", "FS_BIAS.num", "BD_EVAL.num", "BD_SCORE.num", "BD_BIAS.num", "unique.match.y.n", "genome.found.y.n", "basic.genes.y.n","any.cas.genes.y.n" ,"cas.genes", "cas.gene.starts", "cas.gene.stops", "num.of.cas.genes.num")

archaeal.dat <- transform.df(archaeal.dat)
cas9.dat <- transform.df(cas9.dat)
##find the putative cas9 proteins in the list of cas proteins
for(i in 1:nrow(cas9.dat)){
  proteins <- grep(cas9.dat[i, 1], archaeal.dat[,12])
  if(length(proteins) > 0){
  cas9.dat[i, 12] <- F
  }else{
  cas9.dat[i, 12] <- T
  }
}

##check the genomes can be found in the list
for(i in 1:nrow(cas9.dat)){
  gcf <- grep(cas9.dat[i, 3], archaeal.dat[,16])
  if(length(gcf) == 0){
  cas9.dat[i, 13] <- F
  }else{
      cas9.dat[i, 13] <- T
  }
}

##check if basic genes are found in the genomes with cas9 proteins
for(i in 1:nrow(cas9.dat)){
  gcf <- grep(cas9.dat[i, 3], archaeal.dat[,16])
  if(length(gcf) > 0){
    if(archaeal.dat[gcf,3] == T){
        cas9.dat[i, 14] <- T
    }else{
        cas9.dat[i, 14] <- F
    }
  }
}


##check if any other cas genes are found in the genomes with cas9 proteins
for(i in 1:nrow(cas9.dat)){
  gcf <- grep(cas9.dat[i, 3], archaeal.dat[,16])
  if(length(gcf) > 0){
    genes.vector <-  strsplit(archaeal.dat[gcf, 2], ",")
    genes.vector <- genes.vector[[1]][!is.na(genes.vector)]
    if(length(genes.vector)>0){
        cas9.dat[i, 15] <- T
        ##append metadata
        cas9.dat[i, 16] <- archaeal.dat[gcf, 2]
        cas9.dat[i, 17] <- archaeal.dat[gcf, 14]
        cas9.dat[i, 18] <- archaeal.dat[gcf, 15]
        cas9.dat[i, 19] <- length(unique(genes.vector))
        
    }else{
        cas9.dat[i, 15] <- F
    }
  }
}

write.table(cas9.dat, "~/Desktop/Project/identifycasproteins/archaeal_cas9_with_metadata.txt", sep = "\t", quote = F, row.names = F)
```


```{r extract_type_II_data, eval = F, include=T}
##import files of the HMM and rps outputs
archaeal_cas_genes_genomes <- read.table("~/Desktop/Project/identifycasproteins/genomes_with_cas_proteins_archaea_refseq_79_gff_metadata.txt", sep = "\t", comment.char = "", fill = T, header = T)
bacterial_cas_genes_genomes <- read.table("~/Desktop/Project/identifycasproteins/genomes_with_cas_proteins_bacteria_refseq_79_gff_metadata.txt", sep = "\t", comment.char = "", fill = T, header = T)
cas_genes_genomes <- rbind(archaeal_cas_genes_genomes, bacterial_cas_genes_genomes)
cas_genes_genomes <- transform.df(cas_genes_genomes)

cas_genes_genomes <- cas_genes_genomes%>%mutate(subtypes = ifelse(subtypes == ",CAS-I-E, OTHER?", "CAS-I-E", subtypes))%>%mutate(subtypes = ifelse(subtypes == ",CAS-I-F, OTHER?", "CAS-I-F", subtypes))%>%mutate(first.letter.subtypes = substr(subtypes, 1, 1))%>%mutate(subtypes = ifelse(first.letter.subtypes == ",", substr(subtypes, 2, nchar(subtypes)), subtypes))%>%mutate(subtypes.split = strsplit(subtypes, ","))%>%mutate(single.system.y.n = ifelse(substr(subtypes.split, 1, 2) == "c(", F, ifelse(substr(subtypes.split, 1, 2) == "ch", F, T)))%>%select(-subtypes.split)


##extract type II systems and label the taxid data
dat <- cas_genes_genomes
dat <- dat%>%select(genome, subtypes)%>%filter(subtypes != "")
mat <- separate(dat, genome ,c("i1","i2"), sep = "#", remove = F, extra = "drop")%>%select(-i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(i1 = i3)%>%select(-i3, -i2)
mat <- separate(mat, i1 ,c("i2","i3"), sep = "-", remove = F, extra = "merge", fill = "left")%>%mutate(taxid_id = i3)%>%select(-i1, -i3, -i2)

type.II.systems.dat <- mat%>%mutate(`II-A` = ifelse(grepl("CAS-II-A", subtypes), 1, 0))%>%
  mutate(`II-B` = ifelse(grepl("CAS-II-B", subtypes), 1, 0))%>%
  mutate(`II-C` = ifelse(grepl("CAS-II-C", subtypes), 1, 0))%>%
  mutate(keep = `II-A`+`II-B`+`II-C`)%>%
  filter(keep > 0)%>%
  select(-keep, -subtypes)

write.table(type.II.systems.dat, file = "~/Desktop/Project/type.ii.systems.txt", sep = "\t", row.names = F, col.names = T, quote = F)

```

```{r cas_gene_dataframe_setup, eval=F}
##import and set up the cas_genes_genomes variable
archaeal_cas_genes_genomes <- read.table("~/Desktop/Project/identifycasproteins/genomes_with_cas_proteins_archaea_refseq_79_gff_metadata.txt", sep = "\t", comment.char = "", fill = T, header = T)
bacterial_cas_genes_genomes <- read.table("~/Desktop/Project/identifycasproteins/genomes_with_cas_proteins_bacteria_refseq_79_gff_metadata.txt", sep = "\t", comment.char = "", fill = T, header = T)
cas_genes_genomes <- rbind(archaeal_cas_genes_genomes, bacterial_cas_genes_genomes)
cas_genes_genomes <- transform.df(cas_genes_genomes)

cas_genes_genomes <- cas_genes_genomes%>%mutate(subtypes = ifelse(subtypes == ",CAS-I-E, OTHER?", "CAS-I-E", subtypes))%>%mutate(subtypes = ifelse(subtypes == ",CAS-I-F, OTHER?", "CAS-I-F", subtypes))%>%mutate(first.letter.subtypes = substr(subtypes, 1, 1))%>%mutate(subtypes = ifelse(first.letter.subtypes == ",", substr(subtypes, 2, nchar(subtypes)), subtypes))%>%mutate(subtypes.split = strsplit(subtypes, ","))%>%mutate(single.system.y.n = ifelse(substr(subtypes.split, 1, 2) == "c(", F, ifelse(substr(subtypes.split, 1, 2) == "ch", F, T)))%>%select(-subtypes.split)

i <- 1
xx <- cas_genes_genomes[i,]
gene.list <- unlist(strsplit(as.character(xx$genes), ","))
list.length <- length(gene.list)
cas.score.list <- unlist(strsplit(as.character(xx$cas.e.values), ","))
cdd.score.list <- unlist(strsplit(as.character(xx$cdd.e.values), ","))
cas.models.list <- unlist(strsplit(as.character(xx$cas.models), ","))
gff.starts.list <- unlist(strsplit(as.character(xx$gff_starts), ","))
gff.stops.list <- unlist(strsplit(as.character(xx$gff_stops), ","))

if(list.length > length(cdd.score.list)){
  cdd.score.list <- c(cdd.score.list, rep("", list.length - length(cdd.score.list)))
}
if(list.length > length(cas.score.list)){
  cas.score.list <- c(cas.score.list, rep("", list.length - length(cas.score.list)))
}
if(list.length > length(cas.models.list)){
  cas.models.list <- c(cas.models.list, rep("", list.length - length(cas.models.list)))
}
if(list.length > length(gff.starts.list)){
  gff.starts.list <- c(gff.starts.list, rep("", list.length - length(gff.starts.list)))
}
if(list.length > length(gff.stops.list)){
  gff.stops.list <- c(gff.stops.list, rep("", list.length - length(gff.stops.list)))
}

cas.score.list <- cas.score.list[1:list.length]
cdd.score.list <- cdd.score.list[1:list.length]
cas.models.list <- cas.models.list[1:list.length]
gff.starts.list <- gff.starts.list[1:list.length]
gff.stops.list <-gff.stops.list[1:list.length]

GCF.name <- xx$gcf.number
subtype.name <- xx$subtypes
basic.genes.name <- xx$basic.genes.present.y.n
gdf <- data.frame(gene = gene.list, 
                 cas.e.value = cas.score.list, 
                 cdd.e.value = cdd.score.list, 
                 cas.model = cas.models.list, 
                 gene.start = gff.starts.list, 
                 gene.end= gff.stops.list, 
                 GCF_name = rep(GCF.name, list.length), 
                 Subtype = rep(subtype.name, list.length), 
                 basic.genes.present = rep(basic.genes.name, list.length)
)
i <- 2
for(i in 2:nrow(cas_genes_genomes)){
  print(i)
  xx <- cas_genes_genomes[i,]
  gene.list <- unlist(strsplit(as.character(xx$genes), ","))
  list.length <- length(gene.list)
  if(list.length >0){
  cas.score.list <- unlist(strsplit(as.character(xx$cas.e.values), ","))
  cdd.score.list <- unlist(strsplit(as.character(xx$cdd.e.values), ","))
  cas.models.list <- unlist(strsplit(as.character(xx$cas.models), ","))
  gff.starts.list <- unlist(strsplit(as.character(xx$gff_starts), ","))
  gff.stops.list <- unlist(strsplit(as.character(xx$gff_stops), ","))
  
  if(list.length > length(cdd.score.list)){
    cdd.score.list <- c(cdd.score.list, rep("", list.length - length(cdd.score.list)))
  }
  if(list.length > length(cas.score.list)){
    cas.score.list <- c(cas.score.list, rep("", list.length - length(cas.score.list)))
  }
  if(list.length > length(cas.models.list)){
    cas.models.list <- c(cas.models.list, rep("", list.length - length(cas.models.list)))
  }
  if(list.length > length(gff.starts.list)){
    gff.starts.list <- c(gff.starts.list, rep("", list.length - length(gff.starts.list)))
  }
  if(list.length > length(gff.stops.list)){
    gff.stops.list <- c(gff.stops.list, rep("", list.length - length(gff.stops.list)))
  }
  
  cas.score.list <- cas.score.list[1:list.length]
  cdd.score.list <- cdd.score.list[1:list.length]
  cas.models.list <- cas.models.list[1:list.length]
  gff.starts.list <- gff.starts.list[1:list.length]
  gff.stops.list <-gff.stops.list[1:list.length]
  
  GCF.name <- xx$gcf.number
  subtype.name <- xx$subtypes
  basic.genes.name <- xx$basic.genes.present.y.n
  df <- data.frame(gene = gene.list, 
                   cas.e.value = cas.score.list, 
                   cdd.e.value = cdd.score.list, 
                   cas.model = cas.models.list, 
                   gene.start = gff.starts.list, 
                   gene.end= gff.stops.list, 
                   GCF_name = rep(GCF.name, list.length), 
                   Subtype = rep(subtype.name, list.length), 
                   basic.genes.present = rep(basic.genes.name, list.length)
                   )
  gdf <- rbind(gdf, df)
  }
}

write.table(gdf, file = "~/Desktop/genes.dat.txt", col.names = T, row.names = F, quote = F, sep = "\t")

```
CRISPRTarget files I current have for refseq_79:
* Archaea
    * Genbank and refseq phage
    * PHAST
    * IslandViewer
* Bacteria
    * Still Running full search

Using archaeal data to work ont he filtering, genome lengths and setup of the distributions.

Steps to work through:

* Filter the results
    * combine related host genomes to reduce bias from differences in sampling.
    * some reduction will be needed on the protospacer hits where there is more than one match in a genome.
    * find ways to identify related target genomes.
  
* Use genome length information to plot the distances from each spacer (needed for statistical analysis and because the genomes will be treated as circular).
* Decision about using unique arrays or unique genomes is needed.
* Assign initial hit for each target genome.
* Assign the subsequent hits to target and non-target strand.
* Use ks-test to compare the two strands.
* Look at the distribution of all spacers across the genome.
    * Looking to see if clustering is occuring
    * Need a distribution for circular genomes that would be expected by chance (maybe similar to last year's naive distribution).
    
Import the data from the swipe output (does it need to be edited before being imported?). Filter by bit score (>19). Add the spacer id. Separate out the spacer ID to have columns with relavent information (list the columns at some point). Add the genome lengths for the genbank data. Remove lines contain no information or duplicated lines.

```{r swipe_summary, eval = F, include=T}
##import data
arc.genbank <- read.table("~/Desktop/Project/CRISPRClustering/phageq2.archaea.swipe._-1_1", comment.char = "#", header = F, sep = "\t")
bac.genbank <- read.table("~/Desktop/Project/CRISPRClustering/phageq2.bacteria.swipe._-1_1", comment.char = "#", header = F, sep = "\t")
arc.phast <- read.table("~/Desktop/Project/CRISPRClustering/PHASTq.archaea.swipe._-1_1", comment.char = "#", header = F, sep = "\t")
bac.phast <- read.table("~/Desktop/Project/CRISPRClustering/PHASTq.bacteria.swipe._-1_1", comment.char = "#", header = F, sep = "\t")

##name columns
query.colnames <- c("target.acc.", "host.acc.", "%.identity.num", "alignment.length.num", "mismatches.num", "gap.opens.num", "target.start.num", "target.end.num", "host.start.num", "host.end.num", "bit.score.num")
colnames(arc.phast) <- query.colnames
colnames(arc.genbank) <- query.colnames
colnames(bac.phast) <- query.colnames
colnames(bac.genbank) <- query.colnames

##add columns with infomation about search db, Host domain and search type (currently just swipe)
arc.phast <- cbind(arc.phast, search.type = rep("swipe", nrow(arc.phast)), target.db = rep("PHAST", nrow(arc.phast)), Host = rep("Archaea", nrow(arc.phast)))
arc.genbank <- cbind(arc.genbank, search.type = rep("swipe", nrow(arc.genbank)), target.db = rep("Genbank Phage", nrow(arc.genbank)), Host = rep("Archaea", nrow(arc.genbank)))
bac.phast <- cbind(bac.phast, search.type = rep("swipe", nrow(bac.phast)), target.db = rep("PHAST", nrow(bac.phast)), Host = rep("Bacteria", nrow(bac.phast)))
bac.genbank <- cbind(bac.genbank, search.type = rep("swipe", nrow(bac.genbank)), target.db = rep("Genbank Phage", nrow(bac.genbank)), Host = rep("Bacteria", nrow(bac.genbank)))

##import lookup tables for genome length
genbank.lookup <- read.table("~/Desktop/Project/CRISPRClustering/genome_length_lookup_table.txt", header = F)
colnames(genbank.lookup) <- c("Target.Genome", "genome.length.num")
genbank.lookup <- transform.df(genbank.lookup)

##import lookup tables for swipe accessions
archaeal.lookup <- read.table("~/Desktop/Project/CRISPRClustering/phageq.archaea.swipe._-1_1.lookup", fill = T, header = F)
archaeal.lookup <- archaeal.lookup[substr(archaeal.lookup[,1],1,1)==">",]
bacterial.lookup <- read.table("~/Desktop/Project/CRISPRClustering/phageq.bacteria.swipe.lookup", fill = T, header = F)
bacterial.lookup <- bacterial.lookup[substr(bacterial.lookup[,1],1,1)==">",]

##append e.value.num to files
arc.phast <- cbind(arc.phast[,c(1:11)], "evalue.num" =  rep(NA, nrow(arc.phast)), arc.phast[,c(12:ncol(arc.phast))])
arc.genbank <- cbind(arc.genbank[,c(1:11)], "evalue.num" =  rep(NA, nrow(arc.genbank)), arc.genbank[,c(12:ncol(arc.genbank))])
bac.phast <- cbind(bac.phast[,c(1:11)], "evalue.num" =  rep(NA, nrow(bac.phast)), bac.phast[,c(12:ncol(bac.phast))])
bac.genbank <- cbind(bac.genbank[,c(1:11)], "evalue.num" =  rep(NA, nrow(bac.genbank)), bac.genbank[,c(12:ncol(bac.genbank))])
##combine bacteria and archaea
swipe.dat <- rbind(arc.genbank,arc.phast, bac.phast, bac.genbank)
swipe.dat <- unique(swipe.dat)

colnames(swipe.dat)



##filter out low bit scoring numbers
swipe.dat <- swipe.dat%>%filter(bit.score.num > 19)%>%distinct()

swipe.dat <- transform.df(swipe.dat)
bacterial.lookup <- transform.df(bacterial.lookup)
archaeal.lookup <- transform.df(archaeal.lookup)

##edit the column names for the swipe lookup tables so that left_join can be used on the merged file (initially the swipe IDs are the same for archaea and bacteria and need to have the host appended)
x<- rep("Bacteria", nrow(bacterial.lookup))
bacterial.lookup <- cbind(bacterial.lookup, x)
bacterial.lookup[,1]<- with(bacterial.lookup, paste0(V1, x))
bacterial.lookup <- bacterial.lookup[,c(1,2)]
x<- rep("Archaea", nrow(archaeal.lookup))
archaeal.lookup <- cbind(archaeal.lookup, x)
archaeal.lookup[,1]<- with(archaeal.lookup, paste0(V1, x))
archaeal.lookup <- archaeal.lookup[,c(1,2)]

##combine lookup tables
all.lookup <- rbind(bacterial.lookup, archaeal.lookup)
colnames(all.lookup) <- c("host.acc.", "spacer.acc.")
all.lookup[,1] <- with(all.lookup, substr(host.acc., 2, nchar(host.acc.)) )
swipe.dat[,2] <- with(swipe.dat, paste0(host.acc., Host))

##match the swipe IDs with the spacer names that contain all the information
tt <- left_join(swipe.dat, all.lookup, by = "host.acc.")

##replace the column containing swipe ids with the spacer ids.
tt <- tt%>%mutate(host.acc. = spacer.acc.) ##this line is untested. Once it is shown to work, delete the above line that has been commented out.


#"search.type","target.db","Host","","","","","") ##this was sitting here without anything else. I don't know if it should have been deleted or was part of something that should have been kept.

##separate out the information from the spacer id
mat <- separate(tt, host.acc.,c("host","Subtype","Genome","i2","i3"), sep = "\\|", remove = F)


###
mat <- mat%>%mutate(host.target.pair = paste(target.acc., Genome, sep = "_"))
length(unique(mat$target.acc.))
length(unique(mat$Genome))
length(unique(mat$host.target.pair))

i.e <- mat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))






mat <- mat[!is.na(mat[,6]),]
mat <- separate(mat, i2,c("array.start.num","array.stop.num"), sep = "-", remove = T)
mat <- separate(mat, Subtype,c("Type"), sep = "-", remove = F,  extra = "drop")
mat <- separate(mat, i3,c("i5", "i6"), sep = "\\[", remove = T, extra = "drop")
mat <- separate(mat, i5,c("array.num","spacer.num"), sep = "_", remove = T, extra = "merge")
mat <- separate(mat, i6,c("left.flank.num","right.flank.num", "array.score.num", "Taxonomy"), sep = "_", remove = T, extra = "drop")

##select the columns 
swipe.dat <- mat%>%select(target.acc., Genome, host.acc., target.start.num, target.end.num, host.start.num, host.end.num, array.num, spacer.num, Type, Subtype, `bit.score.num`, alignment.length.num, Host, target.db)

##add the genome length to the genbank data using the lookup table
colnames(genbank.lookup) <- c("target.acc." , "genome.length.num")
swipe.dat <- left_join(swipe.dat, genbank.lookup, by = "target.acc.")

##remove lines that contain nno information or are duplicates
swipe.dat <- swipe.dat[!is.na(swipe.dat[,1]),]
swipe.dat.2 <- swipe.dat%>%distinct()


write.table(swipe.dat.2, "~/Desktop/Project/CRISPRClustering/refseq_79.all.swipe_summary", row.names = F, col.names = T, sep = "\t", quote = F)



```


This is now doing the phast lengths, strand data and hits count. There is a filtering step that removes the host-target pairs with a single hit (Currently 575,763 -> 161,800)
```{r phast_lengths, eval = F, include=T}

##import data
targets.dat <- read.table("~/Desktop/Project/CRISPRClustering/refseq_79.all.swipe_summary.no_quotes", comment.char = "", fill = T, sep = "\t", header = T, quote = "")
targets.filtered <- targets.dat[!is.na(targets.dat[,1]),]
targets.filtered <- transform.df(targets.filtered)

nrow(targets.dat%>%filter(Subtype == "I-F")%>%select(Genome, array.num, spacer.num))

##check all genbank genomes have the length #annotated
genbank.targets <- targets.filtered%>%filter(target.db=="Genbank Phage")

##missing genome lengths genbank swipe = genbank.targets.miss
genbank.targets.miss <- genbank.targets%>%filter(is.na(genome.length.num))

###not sure what this is doing
if(nrow(genbank.targets.miss) > 0){
mat <- separate(genbank.targets.miss, host.acc.,c("host","Subtype","Genome","i2","i3"), sep = "\\|", remove = F)
mat <- mat[!is.na(mat[,6]),]
mat <- separate(mat, i2,c("array.start.num","array.stop.num"), sep = "-", remove = T)
mat <- separate(mat, Subtype,c("Type"), sep = "-", remove = F,  extra = "drop")
mat <- separate(mat, i3,c("i5", "i6"), sep = "\\[", remove = T, extra = "drop")

mat <- separate(mat, i5,c("array.num","spacer.num"), sep = "_", remove = T, extra = "merge")

mat <- separate(mat, i6,c("left.flank.num","right.flank.num", "array.score.num", "Taxonomy"), sep = "_", remove = T, extra = "drop")
mat <- mat[,c(1:29)]
mat <- transform.df(mat)
}
#swipe.dat <- mat[,c(1,2,15,16,17,18,19,20,21,22,23,24,25,26,27,6,9,10,8,5,4,7,13,14)]

##get PHAST genome lengths

##get the PHAST results. Take the target.acc. column ans split it in order to get the genome length of the phages.
phast.targets <- targets.filtered%>%filter(target.db == "PHAST")%>%separate(target.acc.,c("i1", "i2"), sep = "_", remove = F,  extra = "drop") %>%separate(i2,c("start.pos.num", "end.pos.num"), sep = "-", remove = T,  extra = "drop")%>%mutate(genome.length.num = abs(as.numeric(end.pos.num) - as.numeric(start.pos.num)))%>%select(-i1, -start.pos.num, -end.pos.num)

##make sure the column names are all the same
colnames(phast.targets) <- colnames(targets.dat)

##combine genbank and phast data
dat <- rbind(genbank.targets, phast.targets)

##remove anything that should not have been introduced
dat <- dat[!is.na(dat[,1]),]
dat <- dat%>%distinct()

##get the strand information for each of the target hit sites
dat <- mutate(dat, strand = ifelse(as.numeric(host.start.num) - as.numeric(host.end.num) > 0, "-", "+"))

##count number of hits to target for each host genome
dat <- dat%>%group_by(Genome, target.acc.)%>%mutate(hits.count.num = n())%>%filter(hits.count.num > 1)

dat <- dat%>%mutate(host.target.pair = paste(target.acc., Genome, sep = "_"))


length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

write.table(dat, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.working", sep = "\t", quote = F, row.names = F, col.names = T)


```


Look at how likely the PPS is to be correct
161,800 hits go into this chunk. A filtering step is done where host target pairs containing only 1 spacer (made it through hits count filter as the spacer matched two spots on the target genome) are removed.
115,622 hits are kept at this step.

The priming protospacer (PPS) which is the oldest spacer in a host target pair as judged by how far along the array it is (larger spacer numbers are older) is assigned (all spacers are assigned based on their order in the array) and scored (based on how confidently they can be assigned as the PPS). Filtering is done at this point to remove host target pairs where it is uncertain which of the spacers is the PPS.
105,695 hits are kept at this step and then taken through to the next chunk.
```{r PPS_score, eval = F, include=T}

##import data
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.working", comment.char = "", fill = T, sep = "\t", header = T, quote = "")
targets.dat <- transform.df(targets.dat)


##Include column containing identifiers for the host.target pairs
targets.dat <- mutate(targets.dat, host.target.pair = paste(Genome, target.acc., sep = "__"))

hits.count <- targets.dat%>%group_by(host.target.pair)%>%summarise(hits.count.x = n())

targets.dat <- left_join(targets.dat, hits.count, by = "host.target.pair")
targets.dat <- targets.dat%>%mutate(hits.count.num = hits.count.x)%>%select(-hits.count.x)%>%filter(hits.count.num != 1)

##get information on the number of spacers in each of the host-target pairs
number.of.spacers <- targets.dat%>%group_by(host.target.pair, array.num, spacer.num)%>%distinct()%>%ungroup()%>%group_by(host.target.pair)%>%summarise(number.of.spacers = n())
targets.dat <- left_join(targets.dat, number.of.spacers, by = "host.target.pair")

##remove host target pairs containing only 1 spacer (made it through hits count filter as the spacer matched two spots on the target genome)
targets.dat <- targets.dat%>%filter(number.of.spacers > 1)%>%select(-number.of.spacers)

hits.count <- targets.dat%>%group_by(host.target.pair)%>%summarise(hits.count.x = n())

targets.dat <- left_join(targets.dat, hits.count, by = "host.target.pair")
targets.dat <- targets.dat%>%mutate(hits.count.num = hits.count.x)%>%select(-hits.count.x)%>%filter(hits.count.num != 1)


##get the number of arrays involved in each host-target pair
number.of.arrays <- targets.dat%>%group_by(host.target.pair)%>%summarise(num.of.arrays.num= length(unique(array.num)))

##add a column containing the number of arrays in each host.target.pair
targets.dat <- left_join(targets.dat, number.of.arrays, by = "host.target.pair")

##score the PPS 5 if the host target pair only has one array.
targets.dat <- mutate(targets.dat, PPS.score.num = ifelse(num.of.arrays.num==1, 5, NA))

##add a column that labels each of the hits based on the age of the spacer (oldest to youngest based on how far along the array that they are found)
targets.dat <- targets.dat%>%group_by(host.target.pair)%>%arrange(-spacer.num)%>%mutate(spacer_order.num = row_number())

##obtain the two oldest spacers and check if these are from the same array
number.of.arrays.first.two.spacers <- targets.dat%>%group_by(host.target.pair)%>%top_n(n = 2, wt = -spacer_order.num)%>%summarise(number.of.arrays.first.two.spacers.num = length(unique(array.num)))

##add the column with the information about whether the two oldest spacers are from the same array
targets.dat <- left_join(targets.dat, number.of.arrays.first.two.spacers, by = "host.target.pair")

##if the two oldest spacers are from the same array then score the PPS 4
targets.dat <- mutate(targets.dat, PPS.score.num = ifelse(number.of.arrays.first.two.spacers.num==1 && num.of.arrays.num > 1, 4, PPS.score.num))
targets.dat <- as.data.frame(targets.dat)

##get the data for the oldest spacers in a new data frame
first.spacer.dat <- targets.dat%>%group_by(host.target.pair)%>%filter(number.of.arrays.first.two.spacers.num == 2)%>%filter(spacer_order.num == 1)

##get the data for the second oldest spacers in a new data frame
second.spacer.dat <- targets.dat%>%group_by(host.target.pair)%>%filter(number.of.arrays.first.two.spacers.num == 2)%>%filter(spacer_order.num == 2)%>%ungroup()%>%select(spacer.num, host.target.pair)
colnames(second.spacer.dat) <- c("spacer.num.2", "host.target.pair")

##combine the data for the first and second spacers so that each of the host.target pairs has a single row with the spacer number of each
first.and.second.spacer.dat <- left_join(first.spacer.dat, second.spacer.dat, by = "host.target.pair")

##add a column where the the percentage difference of the two spacer numbers is recorded
first.and.second.spacer.dat <- mutate(first.and.second.spacer.dat, spacer.diff.num = spacer.num.2/spacer.num)

##if the percentage difference is  then score the PPS 3 and keep the hits with a PPS score 3 or greater
first.and.second.spacer.dat <-  mutate(first.and.second.spacer.dat, PPS.score.num = ifelse(number.of.arrays.first.two.spacers.num==2 && spacer.diff.num < 0.65, 3, 0))%>%filter(PPS.score.num != 0)%>%select(-spacer.diff.num, -spacer.num.2)

##remove the rows from the main data that are contain in the first.and.second.spacer.dat (to be re added with the PPS score data)
targets.dat.pps.4.5 <- targets.dat%>%filter(!is.na(PPS.score.num))
targets.dat.pps.4.5 <- as.data.frame(targets.dat.pps.4.5)

targets.dat.pps.3 <- targets.dat%>%filter(is.na(PPS.score.num))
targets.dat.pps.3 <- as.data.frame(targets.dat.pps.3)

first.and.second.spacer.dat <- as.data.frame(first.and.second.spacer.dat)
first.and.second.spacer.dat <- first.and.second.spacer.dat%>%mutate(x = PPS.score.num)%>%select(host.target.pair, x)
targets.dat.pps.3 <- left_join(targets.dat.pps.3, first.and.second.spacer.dat, by = "host.target.pair")
targets.dat.pps.3 <- targets.dat.pps.3%>%mutate(PPS.score.num = x)%>%select(-x)

##combine the data with a PPS score of 4 and 5 with the data frame with PPS score of 3 (this is now the complete data set again)
targets.dat  <- rbind(targets.dat.pps.4.5, targets.dat.pps.3)

table(targets.dat$PPS.score.num)

##check one more time that all the rows are unique and contain data that is wanted
targets.dat <- targets.dat%>%filter(!is.na(PPS.score.num))%>%distinct()
table(targets.dat$PPS.score.num)



pps.5 <- targets.dat%>%filter(PPS.score.num == 5)
pps.4 <- targets.dat%>%filter(PPS.score.num == 4)
pps.3 <- targets.dat%>%filter(PPS.score.num == 3)

dat <- targets.dat
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))


write.table(targets.dat, file = "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.PPS.scores.working", sep = "\t", quote = F, row.names = F, col.names = T)
```


This chunk takes the dataframe of hits and sorts it into host genome/target genome pairs. The order that the hits information is recorded in is sorted by -spacer number -> host genome -> target genome. Currently the array numbers are being ignored. I don't know if this will matter at a later stage. The first hit is being considered as the spacer with the highest number. 



This chunk adds the target/non-target info.
Input has 98,628
Filter for the spacers that are not the PPS but match the same target site. I don't know if this step is necessary. I might branch off at this point and run two different analyses.
69,497 after filter for same site.
52,376 after host-target pairs that no longer have > 1 hit have been removed
```{r target_non_target_calc, eval = F, include=T}
##import data
targets.dat <- read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.PPS.scores.working", comment.char = "", fill = T, sep = "\t", header = T, quote = "")
targets.dat <- transform.df(targets.dat)

##remove the hits that match the same point as the PPS but are not the PPS.
first.spacer.dat <- targets.dat%>%filter(spacer_order.num == 1)

##get the target start positions for the PPS.
first.spacer.2.columns <- unique(first.spacer.dat%>%select(host.target.pair, target.start.num))

##select the hits that are not PPS
not.first.spacer.dat <- targets.dat%>%filter(spacer_order.num != 1)

##add the target start positions of of the PPS to all the other hits for teh same host.target pair
not.first.spacer.dat <- left_join(not.first.spacer.dat, first.spacer.2.columns, by = "host.target.pair")

##calculate the difference between the PPS start position and other protospacer positions
not.first.spacer.dat <- not.first.spacer.dat%>%mutate(protospacer.distance.num = target.start.num.x - target.start.num.y)

zero.point <- not.first.spacer.dat%>%filter(protospacer.distance.num == 0)
targets.dat <- targets.dat%>%mutate(x = abs(host.end.num - host.start.num))
table(zero.point$Subtype)
table(zero.point$x)
table(targets.dat$x)

hist(targets.dat$x, breaks = 50)
##remove the hits where the site that is matched to is the same as the PPS. This is because it cannot be used for looking at 5' and 3' directions (maybe keep to look at target/non-target information. Is there a biological reason for/against?).
not.first.spacer.dat <- not.first.spacer.dat%>%filter(protospacer.distance.num != 0)
not.first.spacer.dat <- not.first.spacer.dat%>%select(-protospacer.distance.num, - target.start.num.y)
colnames(not.first.spacer.dat) <- colnames(first.spacer.dat)

##combine the PPS and other hits that have been kept
targets.dat <- rbind(first.spacer.dat, not.first.spacer.dat)

dat <- targets.dat
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))


##recount the number of hits, and filter so that only host-target pairs with at least two hits are kept.
targets.dat <- targets.dat%>%group_by(Genome, target.acc.)%>%mutate(hits.count.num = n())%>%filter(hits.count.num > 1)

##add target/non-target strand info. Label all PPS as "t"
targets.dat <- mutate(targets.dat, target.strand = ifelse(spacer_order.num == 1, "t", NA))

##select the PPS rows and the columns containing host-target pair info and strand info
first.spacer.dat <- targets.dat%>%group_by(host.target.pair)%>%filter(spacer_order.num == 1)%>%mutate(i1 = strand)%>%select(host.target.pair, i1)

##add the target info for the PPS to the host-target pairs
targets.dat <- left_join(targets.dat, first.spacer.dat, "host.target.pair")

##add the target/non-target info for each of the hits

targets.dat <- targets.dat%>%mutate(target.strand = ifelse(i1 == "+", ifelse(strand == "+", "t", "n"),ifelse(strand == "+", "n", "t")))%>%select(-i1)

dat <- targets.dat
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

write.table(targets.dat, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.target_strand_2", sep = "\t", quote = F, row.names = F, col.names = T)
```


Distance to the initial protospacer calculation. 
There is no filtering at this step.
```{r protospacer_distances, eval = F, include=T}

##import data
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.target_strand_2", comment.char = "", fill = T, sep = "\t", header = T, quote = "")
targets.dat <- transform.df(targets.dat)

##Get the PPS target start position
first.spacer.dat <- targets.dat%>%group_by(host.target.pair)%>%filter(spacer_order.num == 1)%>%mutate(PPS.start.pos.num = target.start.num)%>%select(host.target.pair, PPS.start.pos.num)

##add the PPS target start position to the other hits in each of the host.target pairs
targets.dat <- left_join(targets.dat, first.spacer.dat, by = "host.target.pair")

##Get linear distance between PPS and protospacer
targets.dat <- targets.dat%>%mutate(i2 = as.numeric(target.start.num) - as.numeric(PPS.start.pos.num))

##identify protospacers that are closer when the genome is treated as linear
targets.dat <- targets.dat%>%mutate(i3 = ifelse(abs(as.numeric(i2)) < as.numeric(genome.length.num)/2, "y","n"))

##Get genome distances from circular genomes
lengeths.greater.than.half <- targets.dat%>%filter(i3 == "n")%>%mutate(protospacer.distance.num = ifelse(i2 < 0 ,as.numeric(i2) + as.numeric(genome.length.num), as.numeric(i2) - as.numeric(genome.length.num)))
lengeths.less.than.half <- targets.dat%>%filter(i3 == "y")%>%mutate(protospacer.distance.num = i2)

##combine the two sets of genomes
targets.dat <- rbind(lengeths.less.than.half, lengeths.greater.than.half)
targets.dat <- targets.dat%>%select(-PPS.start.pos.num, -i2, -i3)

write.table(targets.dat, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.spacer_order_2", sep = "\t", quote = F, row.names = F, col.names = T)

```

Get the direction (5' or 3') from the PPS to the hits.
```{r five_and_3_prime_calc, eval = F, include=T}

##import data
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.spacer_order_2", comment.char = "", fill = T, sep = "\t", header = T, quote = "")
targets.dat <- transform.df(targets.dat)


##get PPS start position and strand
first.spacer.dat <- targets.dat%>%filter(spacer_order.num == 1)%>%select(host.target.pair, strand, target.start.num)%>%mutate(PPS.strand = strand)%>%mutate(PPS.start.pos.num = target.start.num)%>%select(-strand, - target.start.num)

##add the PPS data to the host.target pairs
targets.dat <- left_join(targets.dat, first.spacer.dat, by = "host.target.pair")

##select the data that is not the PPS 
not.PPS.dat <- targets.dat%>%filter(spacer_order.num != 1)

##select the PPS data
first.spacer.dat <- targets.dat%>%filter(spacer_order.num == 1)

##calculate 5' and 3' direction
not.PPS.dat <- not.PPS.dat%>%mutate(five.three.prime.dir = ifelse(PPS.strand == "+", 
                                                         ifelse(target.strand == "t", 
                                                         ifelse(protospacer.distance.num < 0 , "5", "3"), 
                                                         ifelse(protospacer.distance.num < 0 , "3", "5")), 
                                                         ifelse(target.strand == "t", 
                                                         ifelse(protospacer.distance.num < 0 , "3", "5"), 
                                                         ifelse(protospacer.distance.num < 0 , "5", "3"))))
first.spacer.dat <- first.spacer.dat%>%mutate(five.three.prime.dir = "0")

##combine the data
targets.dat <- rbind(first.spacer.dat, not.PPS.dat)

targets.dat%>%group_by(five.three.prime.dir)%>%summarise(counts = n())
targets.dat%>%group_by(target.strand)%>%summarise(counts = n())

targets.dat <- targets.dat%>%mutate(protospacer.distance.num = ifelse(PPS.strand == "-", protospacer.distance.num*(-1), protospacer.distance.num))

write.table(targets.dat, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.5-3_prime_2", sep = "\t", quote = F, row.names = F, col.names = T)

```




### I need to have a proper look through this section to confirm that what it is doing is correct ###


```{r identify_redundant_genomes_2, eval = F, include=T}
##import  data
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.5-3_prime_2", comment.char = "", fill = T, sep = "\t", header = T)

length(unique(targets.dat$target.acc.))
length(unique(targets.dat$Genome))
length(unique(targets.dat$host.target.pair))

##create column containing the target position, the bit score of the match and the spacer order number (should all of these be included?).
targets.dat <- targets.dat%>%mutate(unique.protospacer.host.match =  paste(target.start.num, target.end.num, bit.score.num, spacer_order.num, sep = "_"))%>%select(-target.db)

##combine each of the unique.protospacer.host.matches for each of the host target pairs so that the entire match is summarised in one column.
host.target.pair.summary <- targets.dat%>%group_by(target.acc.,host.target.pair)%>%summarise(all.target.starts = paste(unlist(list(unique.protospacer.host.match)), collapse = ","))
host.target.pair.summary <- as.data.frame(host.target.pair.summary)

##count the number of times each of the host genomes are identical for each of the target genomes.
number.of.duplicate.host.genomes <- host.target.pair.summary%>%group_by(target.acc.,all.target.starts)%>%summarise(duplicate.genomes = n())%>%ungroup()%>%select(-target.acc.)

##add th number of times each of the host genomes occurs to the data set containing the host.target.pair data.
host.target.pair.summary <- left_join(host.target.pair.summary, number.of.duplicate.host.genomes, by = "all.target.starts")
host.target.pair.summary <- host.target.pair.summary%>%ungroup()%>%select(-target.acc.)

##add the duplicate genomes data to the hits data
targets.dat <- left_join(targets.dat, host.target.pair.summary, by = "host.target.pair")
targets.dat <- unique(targets.dat)

##select the hits that are part of duplicated host genomes.
duplicate.host.genomes <- targets.dat%>%filter(duplicate.genomes != 1)
duplicate.host.genomes[,2] <- as.character(duplicate.host.genomes$Genome)


dat <- duplicate.host.genomes
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))


##create a combined 'genome' name containing each of the host genome names that are duplicated for a target genome.
merged.host.genome.names <- duplicate.host.genomes%>%group_by(all.target.starts)%>%summarise(matching.host.genomes = paste(unlist(list(unique(Genome))), collapse = "_merged_"))

##add the matching.host.genomes to the hits data based to label the duplicate genomes by which other genomes are identical.
duplicate.host.genomes <- left_join(duplicate.host.genomes, merged.host.genome.names, by = "all.target.starts")

##relabel the host genome column with the matching.host.genomes column and the the host.target.pair column
duplicate.host.genomes <- duplicate.host.genomes%>%ungroup()%>%mutate(Genome = matching.host.genomes)%>%mutate(host.target.pair = paste(Genome, target.acc., sep = "__"))%>%select(-matching.host.genomes, -host.acc.)
duplicate.host.genomes <- unique(duplicate.host.genomes)


dat <- duplicate.host.genomes
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

##select the hits that where there is no duplicate host genome for a target genome. Further analysis will be done to check these are not similar enough to be called duplcaites.
one.host.genome <- targets.dat%>%filter(duplicate.genomes == 1)%>%select(-host.acc.)
one.host.genome <- unique(one.host.genome)

##combine the now filtered duplicate.host.genomes with the one.host.genome
one.host.genome <- rbind(one.host.genome, duplicate.host.genomes)

##add unique hit information to a column 
one.host.genome <- one.host.genome%>%mutate(unique.protospacer.host.match =  paste(target.acc., target.start.num, target.end.num, sep = "_"))



dat <- one.host.genome
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

##count the number of times a protospacer site is matched
protospacer.counts.dat <- one.host.genome%>%group_by(target.acc., unique.protospacer.host.match)%>%summarise(protospacer.freq.num = n())%>%ungroup()%>%select(-target.acc.)


##add the protospacer counts data
one.host.genome <- left_join(one.host.genome, protospacer.counts.dat, by = "unique.protospacer.host.match")
one.host.genome <- unique(one.host.genome)

##count the number of spacers that are duplicated in each host-target pair
spacer.counts.dat <- one.host.genome%>%filter(protospacer.freq.num > 1)%>%group_by(host.target.pair)%>%summarise(shared.protospacer.num = n())

##add the spacer counts data
one.host.genome <- left_join(one.host.genome, spacer.counts.dat, by ="host.target.pair")

##keep the host-target pairs with no duplicate spacers
no.duplicate.spacer <- one.host.genome%>%filter(is.na(shared.protospacer.num))
no.duplicate.spacer <- unique(no.duplicate.spacer)

dat <- no.duplicate.spacer
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")

length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))


write.table(no.duplicate.spacer, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.no_duplicates", sep = "\t", quote = F, row.names = F, col.names = T)

##keep the host-target pairs with only a single shared protospacer site (this should look at only the cases where the PPS is duplicated initially)
single.duplicate.spacer <- one.host.genome%>%filter(shared.protospacer.num == 1)
write.table(single.duplicate.spacer, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate", sep = "\t", quote = F, row.names = F, col.names = T)


dat <- single.duplicate.spacer
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

##is the PPS the duplicated spacer?
PPS.duplicate.spacer <- single.duplicate.spacer%>%filter(spacer_order.num == 1)%>%mutate(pps.spacer.duplicated = ifelse(protospacer.freq.num > 1, T, F))%>%select(host.target.pair, pps.spacer.duplicated)
single.duplicate.spacer <- left_join(single.duplicate.spacer, PPS.duplicate.spacer, by = "host.target.pair")

##select the spacers that are only duplicated at the PPS
PPS.duplicate.spacer <- single.duplicate.spacer%>%filter(pps.spacer.duplicated == T)

dat <- PPS.duplicate.spacer
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))


write.table(PPS.duplicate.spacer, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate_pps", sep = "\t", quote = F, row.names = F, col.names = T)


##select the remaining spacers
other.spacers <- one.host.genome%>%filter(shared.protospacer.num > 1)
write.table(other.spacers, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.multiple_duplicates", sep = "\t", quote = F, row.names = F, col.names = T)



###
write.table(targets.dat, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.hosts_nr", sep = "\t", quote = F, row.names = F, col.names = T)

```


I am up to here for where I need to continue developing, along with checking previous filter, and other highlighted chunks.


```{r identify_redundant_genomes, eval = F, include=T}

##import data
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.no_duplicates", comment.char = "", fill = T, sep = "\t", header = T)
targets.dat.2 <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate_pps", comment.char = "", fill = T, sep = "\t", header = T)
targets.dat.2 <- targets.dat.2%>%select(-pps.spacer.duplicated)
targets.dat <- rbind(targets.dat, targets.dat.2)

dat <- targets.dat
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))





##create column containing the target position, the bit score of the match and the spacer order number (should all of these be included?).
targets.dat <- targets.dat%>%mutate(unique.spacer.target.match =  paste(spacer.num, array.num, spacer_order.num, Genome, sep = "_"))

##combine each of the unique.spacer.target.matches for each of the host target pairs so that the entire match is summarised in one column.
host.target.pair.summary <- targets.dat%>%group_by(Genome,host.target.pair)%>%summarise(all.host.starts = paste(unlist(list(unique.spacer.target.match)), collapse = ","))
host.target.pair.summary <- as.data.frame(host.target.pair.summary)



##count the number of times each of the host genomes are identical for each of the target genomes.
number.of.duplicate.target.genomes <- host.target.pair.summary%>%group_by(Genome,all.host.starts)%>%summarise(duplicate.targets = n())%>%ungroup()%>%select(-Genome)

##add th number of times each of the host genomes occurs to the data set containing the host.target.pair data.
host.target.pair.summary <- left_join(host.target.pair.summary, number.of.duplicate.target.genomes, by = "all.host.starts")
host.target.pair.summary <- host.target.pair.summary%>%ungroup()%>%select(-Genome)

##add the duplicate genomes data to the hits data
targets.dat <- left_join(targets.dat, host.target.pair.summary, by = "host.target.pair")
targets.dat <- unique(targets.dat)




####

##select the duplicated genomes
duplicate.target.genomes <- targets.dat%>%filter(duplicate.targets > 1)
##create a combined 'genome' name containing each of the host genome names that are duplicated for a target genome.
merged.target.genome.names <- duplicate.target.genomes%>%group_by(all.host.starts)%>%summarise(matching.target.genomes = paste(unlist(list(unique(target.acc.))), collapse = "_merged_"))

##add the matching.host.genomes to the hits data based to label the duplicate genomes by which other genomes are identical.
duplicate.target.genomes <- left_join(duplicate.target.genomes, merged.target.genome.names, by = "all.host.starts")

##relabel the host genome column with the matching.host.genomes column and the the host.target.pair column

tt <- duplicate.target.genomes%>%group_by(matching.target.genomes, spacer.num, array.num)%>%summarise(mean.protospacer.distance = mean(protospacer.distance.num))
tt <- tt%>%mutate(distance.tolerance = 0.40*mean.protospacer.distance)%>%mutate(merged.spacer.id = paste(matching.target.genomes, spacer.num, array.num, sep = "_"))%>%ungroup()%>%select(-matching.target.genomes, -spacer.num, -array.num)
duplicate.target.genomes <- duplicate.target.genomes%>%mutate(merged.spacer.id = paste(matching.target.genomes, spacer.num, array.num, sep = "_"))
duplicate.target.genomes <- left_join(duplicate.target.genomes, tt, by = "merged.spacer.id")
duplicate.target.genomes <- duplicate.target.genomes%>%mutate(protospacer.distance.num.2 = ifelse(abs(protospacer.distance.num -  mean.protospacer.distance) < abs(distance.tolerance), mean.protospacer.distance, protospacer.distance.num ))%>%mutate(distance.is.averaged = ifelse(protospacer.distance.num.2 == mean.protospacer.distance, T, F))

table(duplicate.target.genomes$distance.is.averaged)

#tt <- duplicate.target.genomes%>%group_by(matching.target.names)%>%


duplicate.target.genomes <- duplicate.target.genomes%>%ungroup()%>%mutate(target.acc. = matching.target.genomes)%>%mutate(host.target.pair = paste(Genome, target.acc., sep = "__"))%>%mutate(protospacer.distance.num = protospacer.distance.num.2)%>%select(-matching.target.genomes, -target.start.num, -target.end.num, -protospacer.distance.num.2, -genome.length.num, -unique.protospacer.host.match, -all.target.starts, -host.start.num,-host.end.num, -strand, -bit.score.num, -alignment.length.num, -shared.protospacer.num, -PPS.start.pos.num, -PPS.score.num, -distance.is.averaged, -distance.tolerance, -mean.protospacer.distance, -merged.spacer.id, -hits.count.num, -num.of.arrays.num, -number.of.arrays.first.two.spacers.num)

duplicate.target.genomes <- duplicate.target.genomes%>%mutate(spacer_order.num = ifelse(spacer_order.num == 1, 1, 2))
duplicate.target.genomes <- unique(duplicate.target.genomes)

####



##select the target genomes that are unique
one.target.genome <- targets.dat%>%filter(duplicate.targets == 1)

one.target.genome <- one.target.genome%>%select( -target.start.num, -target.end.num, -genome.length.num, -unique.protospacer.host.match, -all.target.starts, -host.start.num,-host.end.num, -strand, -bit.score.num, -alignment.length.num, -shared.protospacer.num, -PPS.start.pos.num, -PPS.score.num, -hits.count.num, -num.of.arrays.num, -number.of.arrays.first.two.spacers.num)

one.target.genome <- rbind(one.target.genome, duplicate.target.genomes)

##count the number of times a protospacer site is matched
spacer.counts.dat <- one.target.genome%>%group_by(Genome, unique.spacer.target.match)%>%summarise(spacer.freq.num = n())%>%ungroup()%>%select(-Genome)

##add the protospacer counts data
one.target.genome <- left_join(one.target.genome, spacer.counts.dat, by = "unique.spacer.target.match")
one.target.genome <- unique(one.target.genome)

##count the number of spacers that are duplicated in each host-target pair
spacer.counts.dat <- one.target.genome%>%filter(spacer.freq.num > 1)%>%group_by(host.target.pair)%>%summarise(shared.spacer.num = n())

##add the spacer counts data
one.target.genome <- left_join(one.target.genome, spacer.counts.dat, by ="host.target.pair")

##keep the host-target pairs with no duplicate spacers
no.duplicate.spacer <- one.target.genome%>%filter(is.na(shared.spacer.num))
write.table(no.duplicate.spacer, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_target_genome", sep = "\t", quote = F, row.names = F, col.names = T)



##keep the host-target pairs with only a single shared protospacer site (this should look at only the cases where the PPS is duplicated initially)
single.duplicate.protospacer <- one.target.genome%>%filter(shared.spacer.num == 1)
write.table(single.duplicate.protospacer, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate", sep = "\t", quote = F, row.names = F, col.names = T)

##is the PPS the duplicated protospacer?
PPS.duplicate.protospacer <- single.duplicate.protospacer%>%filter(spacer_order.num == 1)%>%mutate(pps.protospacer.duplicated = ifelse(spacer.freq.num > 1, T, F))%>%select(host.target.pair, pps.protospacer.duplicated)
single.duplicate.protospacer <- left_join(single.duplicate.protospacer, PPS.duplicate.protospacer, by = "host.target.pair")

##select the spacers that are only duplicated at the PPS
PPS.duplicate.protospacer <- single.duplicate.protospacer%>%filter(pps.protospacer.duplicated == T)
write.table(PPS.duplicate.protospacer, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate_pps_target", sep = "\t", quote = F, row.names = F, col.names = T)




##############################
```

```{r identify_redundant_genomes_old}
##select the duplicated genomes
duplicate.target.genomes <- targets.dat%>%filter(duplicate.targets > 1)
##create a combined 'genome' name containing each of the host genome names that are duplicated for a target genome.
merged.target.genome.names <- duplicate.target.genomes%>%group_by(all.host.starts)%>%summarise(matching.target.genomes = paste(unlist(list(unique(target.acc.))), collapse = "_merged_"))

##add the matching.host.genomes to the hits data based to label the duplicate genomes by which other genomes are identical.
duplicate.target.genomes <- left_join(duplicate.target.genomes, merged.target.genome.names, by = "all.host.starts")

##relabel the host genome column with the matching.host.genomes column and the the host.target.pair column

tt <- duplicate.target.genomes%>%group_by(matching.target.genomes, spacer.num, array.num)%>%summarise(mean.protospacer.distance = mean(protospacer.distance.num))
tt <- tt%>%mutate(distance.tolerance = 0.40*mean.protospacer.distance)%>%mutate(merged.spacer.id = paste(matching.target.genomes, spacer.num, array.num, sep = "_"))%>%ungroup()%>%select(-matching.target.genomes, -spacer.num, -array.num)
duplicate.target.genomes <- duplicate.target.genomes%>%mutate(merged.spacer.id = paste(matching.target.genomes, spacer.num, array.num, sep = "_"))
duplicate.target.genomes <- left_join(duplicate.target.genomes, tt, by = "merged.spacer.id")
duplicate.target.genomes <- duplicate.target.genomes%>%mutate(protospacer.distance.num.2 = ifelse(abs(protospacer.distance.num -  mean.protospacer.distance) < abs(distance.tolerance), mean.protospacer.distance, protospacer.distance.num ))%>%mutate(distance.is.averaged = ifelse(protospacer.distance.num.2 == mean.protospacer.distance, T, F))

table(duplicate.target.genomes$distance.is.averaged)

#tt <- duplicate.target.genomes%>%group_by(matching.target.names)%>%


duplicate.target.genomes <- duplicate.target.genomes%>%ungroup()%>%mutate(target.acc. = matching.target.genomes)%>%mutate(host.target.pair = paste(Genome, target.acc., sep = "__"))%>%select(-matching.target.genomes, -target.start.num, -target.end.num, -protospacer.distance.num,                                                                                                                                -genome.length.num, -unique.protospacer.host.match, -all.target.starts, -host.start.num,-host.end.num, -strand, -bit.score.num, -alignment.length.num)
duplicate.target.genomes <- unique(duplicate.target.genomes)


##count the number of times a protospacer site is matched
spacer.counts.dat <- duplicate.target.genomes%>%group_by(Genome, unique.spacer.target.match)%>%summarise(spacer.freq.num = n())%>%ungroup()%>%select(-Genome)

##add the protospacer counts data
duplicate.target.genomes <- left_join(duplicate.target.genomes, spacer.counts.dat, by = "unique.spacer.target.match")
duplicate.target.genomes <- unique(duplicate.target.genomes)

##count the number of spacers that are duplicated in each host-target pair
spacer.counts.dat <- duplicate.target.genomes%>%filter(spacer.freq.num > 1)%>%group_by(host.target.pair)%>%summarise(shared.spacer.num = n())

##add the spacer counts data
duplicate.target.genomes <- left_join(duplicate.target.genomes, spacer.counts.dat, by ="host.target.pair")

##keep the host-target pairs with no duplicate spacers
no.duplicate.spacer.reduced <- duplicate.target.genomes%>%filter(is.na(shared.spacer.num))


##add the previous data to the filtered duplicate data
one.target.genome.reduced <- one.target.genome%>%select( -target.start.num, -target.end.num, -protospacer.distance.num,
                                                        -genome.length.num, -unique.protospacer.host.match, -all.target.starts, -host.start.num, 
                                                        -host.end.num, -strand, -bit.score.num, -alignment.length.num)

no.duplicate.spacer.reduced <- rbind(no.duplicate.spacer.reduced, one.target.genome.reduced)

write.table(no.duplicate.spacer.reduced, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_target_genome.reduced", sep = "\t", quote = F, row.names = F, col.names = T)



##keep the host-target pairs with only a single shared protospacer site (this should look at only the cases where the PPS is duplicated initially)
single.duplicate.protospacer.reduced <- duplicate.target.genomes%>%filter(shared.spacer.num == 1)
write.table(single.duplicate.protospacer.reduced, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate.reduced", sep = "\t", quote = F, row.names = F, col.names = T)

##is the PPS the duplicated protospacer?
PPS.duplicate.protospacer.reduced <- single.duplicate.protospacer.reduced%>%filter(spacer_order.num == 1)%>%mutate(pps.protospacer.duplicated = ifelse(spacer.freq.num > 1, T, F))%>%select(host.target.pair, pps.protospacer.duplicated)
single.duplicate.protospacer.reduced <- left_join(single.duplicate.protospacer.reduced, PPS.duplicate.protospacer.reduced, by = "host.target.pair")

##select the spacers that are only duplicated at the PPS
PPS.duplicate.protospacer.reduced <- single.duplicate.protospacer.reduced%>%filter(pps.protospacer.duplicated == T)

##add the previous data to the filtered duplicate data
PPS.duplicate.protospacer <- PPS.duplicate.protospacer%>%select( -target.start.num, -target.end.num, -protospacer.distance.num,
                                                         -genome.length.num, -unique.protospacer.host.match, -all.target.starts, -host.start.num, 
                                                         -host.end.num, -strand, -bit.score.num, -alignment.length.num)

PPS.duplicate.protospacer.reduced <- rbind(PPS.duplicate.protospacer.reduced, PPS.duplicate.protospacer)


write.table(PPS.duplicate.protospacer.reduced, "~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate_pps_target.reduced", sep = "\t", quote = F, row.names = F, col.names = T)

####################

```


The protospacer.distance.num variable is changed in this section so that rather than being related to the absolute numbers, it is treated the same way that the quadrant analysis is done. This is done in order to keep consistent the target and non-target strands and 5' or 3' direction. 

These are the rules for the distances.

target strand + 5' = +
target strand + 3' = -

non-target strand + 5' = -
non-target strand + 3' = +

```{r plots_and_tables_distribution_analysis, results='asis'}
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_target_genome.test", comment.char = "", fill = T, sep = "\t", header = T)
targets.dat <- transform.df(targets.dat)

targets.dat.2 <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate.protospacer.test", comment.char = "", fill = T, sep = "\t", header = T)
targets.dat.2 <- transform.df(targets.dat.2)
#targets.dat.2 <- targets.dat.2%>%select(-pps.protospacer.duplicated)

colnames(targets.dat)
colnames(targets.dat.2)

targets.dat <- rbind(targets.dat, targets.dat.2)

table(targets.dat$Subtype)


dat <- targets.dat
length(unique(dat$target.acc.))
length(unique(dat$Genome))
length(unique(dat$host.target.pair))

i.e <- dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

i.e <- i.e%>%filter(spacer_order.num != 1)

#df <- targets.dat%>%select(Subtype)%>%distinct()
df <- cbind(targets.dat%>%select(Subtype, host.target.pair)%>%distinct()%>%group_by(Subtype)%>%summarise(Host.target.pairs = n()),targets.dat%>%select(Subtype, target.acc.)%>%distinct()%>%group_by(Subtype)%>%summarise(Target.genomes = n())%>%select(-Subtype),targets.dat%>%select(Subtype, Genome)%>%distinct()%>%group_by(Subtype)%>%summarise(Host.genomes = n())%>%select(-Subtype),targets.dat%>%group_by(Subtype)%>%summarise(Hits = n())%>%select(-Subtype))
df <- rbind(df, NA)
df[17,1] <- "All"
df[17,2] <- sum(df[c(1:16),2])
df[17,3] <- sum(df[c(1:16),3])
df[17,4] <- sum(df[c(1:16),4])
df[17,5] <- sum(df[c(1:16),5])


###
#targets.dat <- targets.dat.keep
#targets.dat <- targets.dat%>%filter(spacers.ambig.num == 1)

targets.dat <- targets.dat%>%filter(spacer_order.num > 1)

table(targets.dat$Subtype)


##adjust the direction of the distance to protospacer based on the strand and 5/3 prime direction so that each of the individual genomes are the same.
targets.dat <- targets.dat%>%mutate(protospacer.distance.num = ifelse(target.strand == "t", ifelse(five.three.prime.dir == 5, abs(protospacer.distance.num), -1*abs(protospacer.distance.num)), ifelse(five.three.prime.dir == 3, abs(protospacer.distance.num), -1*abs(protospacer.distance.num))))

xx <- targets.dat%>%group_by(protospacer.distance.num)%>%summarise(counts = n())
table(targets.dat$target.strand)
table(targets.dat$five.three.prime.dir)
table(targets.dat$bit.score.num)
table(targets.dat$hits.count.num)
table(targets.dat$PPS.score.num)


##Testing current setup
#targets.dat.2 <- targets.dat
#targets.dat <- targets.dat.2
#tt <- targets.dat
#targets.dat <- tt

nrow(targets.dat%>%filter(protospacer.distance.num == 0))

#targets.dat <- targets.dat%>%filter(spacer_order.num != 1)

targets.dat <- targets.dat%>%filter(protospacer.distance.num > -15000)%>%filter(protospacer.distance.num < 15000)

table(targets.dat$Subtype)

i.e <- targets.dat%>%filter(Subtype == "I-E")
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))

i.e <- i.e%>%filter(spacer_order.num != 1)

length(unique(targets.dat$target.acc.))
length(unique(targets.dat$Genome))
length(unique(targets.dat$target.start.num))
nrow(unique(targets.dat%>%ungroup%>%select(Genome, spacer.num, array.num)))

table(targets.dat$Subtype)

trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t",]
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n",]
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Subtype == "I-A")#[nt[,21]=="I-A",]
trgt <- trgt%>%filter(Subtype == "I-A")#[trgt[,21]=="I-A",]
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

#p=ggplot(dat, aes(x=a)) + geom_histogram(aes(y=..density..),breaks = seq(4,20,by=2))+xlab("Required Solving Time")


ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  #facet_wrap(~protospacer.distance.num)
  
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-A hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  coord_cartesian(ylim = c(-6, 6)) +
  theme_bw()

#pg <- ggplot_build(p)
#x.1 <- pg$data[[1]]
#x.2 <- pg$data[[2]]

table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-A second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()


trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t",]
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n",]
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Subtype =="I-B")#
trgt <- trgt%>%filter(Subtype =="I-B")#
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-B hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()



table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-B second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()


trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t")#
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Subtype =="I-C")#
trgt <- trgt%>%filter(Subtype =="I-C")#
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-C hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()



table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-C second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()


trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t")#
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Subtype =="I-D")#
trgt <- trgt%>%filter(Subtype =="I-D")#
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-D hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()



table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-D second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()

trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t")#
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Subtype =="I-E")#
trgt <- trgt%>%filter(Subtype =="I-E")#
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-E hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()

length(unique(targets.dat.n0$host.target.pair))
length(unique(targets.dat.n0$Genome))
length(unique(targets.dat.n0$target.acc.))

table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)

ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-E second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()

trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t")#
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Subtype =="I-F")#
trgt <- trgt%>%filter(Subtype =="I-F")#
targets.dat.n0 <- rbind(trgt, nt)

length(unique(targets.dat.n0$host.target.pair))
length(unique(targets.dat.n0$Genome))
length(unique(targets.dat.n0$target.acc.))

x <- unique(targets.dat.n0$Genome)

#tt <- targets.dat.n0%>%filter(Genome == x[1]|| x[3] )# || "NZ_JEYC01000034-Acinetobacter_baumannii_83444_ab83444.contig.33,_whole_genome" || "NZ_ASFN01000054-Acinetobacter_baumannii_TG22148_AB_TG22148_55,_whole_genome_shotgun_merged_NZ_ASFN01000020-Acinetobacter_baumannii_TG22148_AB_TG22148_21,_whole_genome_shotgun_merged_NC_011586-Acinetobacter_baumannii_AB0057,_complete_genome." || "NZ_AMSW01000105-Acinetobacter_baumannii_Naval-82_ctg7180000004786,_whole_genome" || "NZ_AMSW01000169-Acinetobacter_baumannii_Naval-82_ctg7180000004765,_whole_genome_merged_NZ_JEYC01000164-Acinetobacter_baumannii_83444_ab83444.contig.163,_whole_genome")

#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

#targets.dat.n0 <- tt

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-F hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()



table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Subtype I-F second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()

trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t")#
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Type=="II")
trgt <- trgt%>%filter(Type=="II")
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Type II hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()

tt <- targets.dat%>%mutate(distance.prop = target.start.num/genome.length.num)%>%filter(spacer_order.num == 1)%>%filter(Type == "II")

#ggplot() +   geom_histogram(data = tt, binwidth = .01, aes(distance.prop))

#table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Type II second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()

trgt <- targets.dat%>%filter(target.strand == "t")#[targets.dat.n0[,28]=="t")#
nt <- targets.dat%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nrow(trgt)/nrow(nt)
nt <- nt%>%filter(Type=="III")
trgt <- trgt%>%filter(Type=="III")
targets.dat.n0 <- rbind(trgt, nt)
#ks.test(pos.i.f[,30], neg.i.f[,30])
#hist(pos.i.f[,30], breaks = 50)
#hist(neg.i.f[,30], breaks = 50)

ggplot() + 
  geom_histogram(data=subset(targets.dat.n0, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat.n0, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Type III hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()



#table(targets.dat.n0$spacer_order.num)

targets.dat_2 <- targets.dat.n0%>%filter(spacer_order.num == 2)
ggplot() + 
  geom_histogram(data=subset(targets.dat_2, target.strand=="t"), binwidth = 1000, aes(protospacer.distance.num, fill = "t", y= ..count..)) +
  geom_histogram(data=subset(targets.dat_2, target.strand=="n"), binwidth = 1000, aes(protospacer.distance.num, fill = "n", y= -..count..)) +
  scale_fill_hue("Group") +
  ggtitle(paste("Distribution of Type III second hit (", nrow(targets.dat_2), " hits)", sep = "")) +
  labs(x="Distance from oldest protospacer (nucleotides)",y="Number of hits") +
  theme_bw()
```

```{r quadrant_analysis, results='asis'}
targets.dat <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_target_genome.test", comment.char = "", fill = T, sep = "\t", header = T)
targets.dat <- transform.df(targets.dat)

targets.dat.2 <-  read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.one_duplicate.protospacer.test", comment.char = "", fill = T, sep = "\t", header = T)
targets.dat.2 <- transform.df(targets.dat.2)
targets.dat.2 <- targets.dat.2%>%select(-pps.protospacer.duplicated)

colnames(targets.dat)
colnames(targets.dat.2)

targets.dat <- rbind(targets.dat, targets.dat.2)

#targets.dat <- targets.dat%>%filter(spacer_order.num > 1)


#targets.dat <- targets.dat%>%filter(spacer_order.num == 2)

length(unique(targets.dat$target.acc.))
length(unique(targets.dat$Genome))
length(unique(targets.dat$host.target.pair))

table(targets.dat$Subtype)

i.e <- targets.dat%>%filter(Subtype == "I-F")#%>%filter(spacer_order.num != 1)
length(unique(i.e$target.acc.))
length(unique(i.e$Genome))
length(unique(i.e$host.target.pair))


targets.dat.n0 <- targets.dat%>%filter(spacer_order.num > 1)%>%filter(spacer_order.num == 2)#%>%filter(nr.genome.to.keep == "y")

nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype == "I-F")#[nt[,21]=="I-F")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype == "I-F")#[trgt[,21]=="I-F")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 

tt <- rbind(trgt, nt)

q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtype I-F data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 




#targets.dat <- targets.dat.keep
targets.dat.n0 <- targets.dat%>%filter(spacer_order.num > 1)%>%filter(spacer_order.num == 2)#%>%filter(nr.genome.to.keep == "y")
#targets.dat.n0 <- targets.dat.n0%>%filter(spacer_order.num == 2)#%>%filter(protospacer.distance.num > -1500)%>%filter(protospacer.distance.num < 1500)

i.f <- targets.dat.n0%>%filter(Subtype == "I-F")
#targets.dat.n0 <- targets.dat%>%filter(protospacer.distance.num != 0) #[targets.dat[,30]!=0,]
nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype == "I-A")#[nt[,21]=="I-A")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype == "I-A")#[trgt[,21]=="I-A")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtype I-A data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype =="I-B")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype =="I-B")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtype I-B data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype =="I-C")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype =="I-C")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtyoe I-C data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype =="I-D")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype =="I-D")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtype I-D data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype == "I-E")#[nt[,21]=="I-E")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype == "I-E")#[trgt[,21]=="I-E")#



t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtype I-E data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
#nt <- nt[nt[,20]=="III")#
nt <- nt%>%filter(Subtype == "I-F")#[nt[,21]=="I-F")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
#trgt <- trgt[trgt[,20]=="III")#
trgt <- trgt%>%filter(Subtype == "I-F")#[trgt[,21]=="I-F")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Subtype I-F data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nt <- nt%>%filter(Type=="II")
#nt <- nt%>%filter(Subtype =="I-A")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
trgt <- trgt%>%filter(Type=="II")
#trgt <- trgt%>%filter(Subtype =="I-A")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Type II data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


nt <- targets.dat.n0%>%filter(target.strand == "n")#[targets.dat.n0[,28]=="n")#
nt <- nt%>%filter(Type=="III")
#nt <- nt%>%filter(Subtype =="I-A")#
nt.3.p <- nt%>%filter(five.three.prime.dir == 3) #[nt[,36]==3,]
nt.5.p <- nt%>%filter(five.three.prime.dir == 5) #nt[nt[,36]==5,]
trgt <- targets.dat.n0%>%filter(target.strand == "t") #targets.dat.n0[targets.dat.n0[,28]=="t")#
trgt <- trgt%>%filter(Type=="III")
#trgt <- trgt%>%filter(Subtype =="I-A")#

t.3.p <- trgt%>%filter(five.three.prime.dir == 3) 
t.5.p <- trgt%>%filter(five.three.prime.dir == 5) 


q.table <- data.frame(PPS.quadrant.num = c(-10,-5,-5,5,5,10), freq.num = c(0,nrow(t.5.p), -nrow(nt.5.p), nrow(t.3.p), -nrow(nt.3.p),0), strand = c("Target","Target", "Non-Target","Target", "Non-Target","Target"))

q.tab.2 <- data.frame(t = c(nrow(t.5.p), nrow(t.3.p)), n = c(nrow(nt.5.p), nrow(nt.3.p)))
rownames(q.tab.2) <- c("5'", "3'")

x <- fisher.test(q.tab.2)
ggplot(q.table, aes(x=PPS.quadrant.num, y=freq.num, fill = strand)) + 
  geom_bar(stat="identity", position="identity") +
  ggtitle("Proportion of newly targeted protospacers relative to the PPS (Type III data)") +
  labs(x="Left is the 5' direction and right is the 3' direction",y="Strand") +
  theme_bw() +
  annotate("text", x=-1, y=7, label= paste("p-value =", round(x$p.value, 4)))# 0.71") 


```

```{r tests, echo=F, eval=F}
table(targets.dat$target.strand)
tt.1 <- targets.dat%>%filter(protospacer.distance.num < -1000)
tt.2 <- targets.dat%>%filter(protospacer.distance.num > 1000)
tt <- rbind(tt.1, tt.2)
table(tt$target.strand)


tt <- targets.dat%>%select(target.acc., Genome, spacer.num, array.num, protospacer.distance.num, target.strand)%>%distinct()
tt <- targets.dat%>%filter(spacer_order.num == 2)
tt.1 <- tt%>%filter(protospacer.distance.num < -20)
tt.2 <- tt%>%filter(protospacer.distance.num > 20)
tt <- rbind(tt.1, tt.2)
table(tt$target.strand)
############
genes.summary <- read.table("~/Desktop/refseq_79.gene_predictions.full.gene_descriptions.txt", sep = "\t", comment.char = "", fill = T, header = T)

crispr_models_data <- read.table("~/Desktop/Project/identifycasproteins/genes_and_systems_summary.txt", as.is = T, sep = '\t', comment.char = "", header = T)
crispr_models_data <- crispr_models_data[crispr_models_data[,3]==T,]
crispr_models_data <- crispr_models_data[crispr_models_data[,4]==T,]

tt <- crispr_models_data[,c(1,2,3)]



colnames(tt) <- c("gene.name", "System", "keep")

aa <- with(tt, paste0(gene.name, "_",System))

tt <- cbind(tt, aa)

tt.1 <- left_join(genes.summary, tt, by = "gene.name")

tt.1 <- tt.1%>%filter(keep == T)

ggplot(tt.1, aes(x=correct.match, y=Freq)) + 
  geom_bar(stat="identity", position="identity") 


tt <- tt.1%>%filter(correct.match == "p")
ggplot(tt, aes(x=aa, y=Freq)) + 
  theme(axis.text.x=element_text(angle=90,hjust=1)) +
  geom_bar(stat="identity", position="identity") +
  ggtitle("Number of times signature genes were found by Makarova models in data that were missed by refseq annotation") +
  labs(x="Gene name",y="Number of times the gene was found") 

tt.2 <- tt%>%filter(gene.name != "csa3")
tt.2 <- tt.2%>%filter(gene.name != "cas3")
tt.2 <- tt.2%>%filter(gene.name != "cas7")

ggplot(tt.2, aes(x=gene.name, y=Freq)) + 
  theme(axis.text.x=element_text(angle=90,hjust=1)) +
  geom_bar(stat="identity", position="identity")

ggplot(genes.summary, aes(x=gene.name, y=Freq)) + 
  theme(axis.text.x=element_text(angle=90,hjust=1)) +
  geom_bar(stat="identity", position="identity")


shuffled.dat <- read.table("~/Desktop/PHAST.Shuffled.bacteria.swipe._-1_1", sep = "\t", comment.char = "#", fill = T, header = F)
swipe.targets <- read.table("~/Desktop/Project/CRISPRClustering/refseq_79.all.swipe_summary", comment.char = "", fill = T, sep = "\t", header = T)
colnames(shuffled.dat) <- c("target.acc.", "host.acc.", "identity.num", "alignment.length.num", "mismatches.num", "gap.opens.num", "target.start.num", "target.end.num", "host.start.num", "host.end.num", "bit.score.num")
#colnames(swipe.targets) <- c("target.acc.", "host.acc.", "identity.num", "alignment.length.num", "mismatches.num", "gap.opens.num", "target.start.num", "target.end.num", "host.start.num", "host.end.num", "bit score.num")

dat <- swipe.targets[,c(1:17)]%>%filter(Host == "Bacteria")%>%filter(target.db == "PHAST")
dat <- dat[,c(1:11)]
shuffled.dat <- shuffled.dat[shuffled.dat[,4]>20,]
shuffled.dat <- shuffled.dat[shuffled.dat[,3] >67,]
shuffled.dat <- transform.df(shuffled.dat)

dat <- dat[dat[,4]>20,]
dat <- dat[dat[,3] >67,]

shuffled.dat <- mutate(shuffled.dat, data.type = "shuffled")
dat <- mutate(dat, data.type = "actual")
colnames(dat) <- colnames(shuffled.dat)

colnames(dat) <- c("target.acc.", "host.acc.", "identity.num", "alignment.length.num", "mismatches.num", "gap.opens.num", "target.start.num", "target.end.num", "host.start.num", "host.end.num", "bit.score.num", "data.type")
dat <- rbind(dat, shuffled.dat)

#ggplot(shuffled.dat, aes(identity.num)) + 
 # geom_histogram(alpha = 0.2, binwidth = 1) +
#  theme_bw()

#ggplot(dat, aes(identity.num, fill = data.type)) + 
#  geom_histogram(alpha = 0.2, binwidth = 1) +
#  theme_bw()

#ggplot(dat, aes(alignment.length.num, fill = data.type)) + 
 # geom_histogram(alpha = 0.2, binwidth = 1) +
#  theme_bw()

#ggplot(dat, aes(mismatches.num, fill = data.type)) + 
#  geom_histogram(alpha = 0.2, binwidth = 1) +
#  theme_bw()

dat.1 <- dat%>%filter(bit.score.num < 50)

ggplot(dat.1, aes(bit.score.num, fill = data.type)) + 
  geom_histogram(alpha = 0.2, binwidth = 1)# +
  theme_bw() +
  ggtitle(paste("Distribution of bit scores for shuffled and PHAST bacterial hits (", nrow(targets.dat.n0), " hits)", sep = "")) +
  labs(x="Score for each hit",y="Number of hits") 


df <- data.frame(phast.num = rep(NA, 50), shuffled.num = rep(NA, 50), error.rate.num = rep(NA, 50), bit.score.cut.off.num = rep(NA, 50))
i <- 15
for(i in 1:50){
tmp <- dat%>%filter(bit.score.num > i)
x <- as.data.frame(table(tmp$data.type))
df[i, 1] <- x[1, 2]
df[i, 2] <- x[2,2]
df[i, 3] <- round((df[i, 2]/df[i, 1])*100, 3)
df[i, 4] <- i

}
plot(x = df$bit.score.cut.off.num, y = df$error.rate.num, xlab = "Score for each hit", ylab = "Error rate (percentage of the data that is there by chance)", main = "Error rate of the swipe search results based on bit scores (Comparison of shuffled PHAST database and actual PHAST database)")
plot(x = df$bit.score.cut.off.num, y = df$phast.num)


dat.2 <- dat.1%>%filter(bit.score.num > 19)
ggplot(dat.2, aes(bit.score.num, fill = data.type)) + 
  geom_histogram(alpha = 0.2, binwidth = 1) +
  theme_bw()


hist(shuffled.dat$`%.identity.num`)



```


Analysis of data (not part of the main throughput)
Importing swipe data and looking at the distribution of e-values and alignment lengths.
```{r protospacer_analysis_1, eval =F}

##import data
swipe.targets <- read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.strand.filtered.working", comment.char = "", fill = T, sep = "\t", header = T)
swipe.targets <- swipe.targets[!is.na(swipe.targets[,1]),]
swipe.targets <- transform.df(swipe.targets)
swipe.targets <- swipe.targets[order(swipe.targets[,1]),]
swipe.targets <- swipe.targets[order(swipe.targets[,2]),]
##check that all the target genome lengths have been obtained
length(unique(swipe.targets[is.na(swipe.targets[,25]), 1]))

##genomes without lengths recoreded
unique(swipe.targets[is.na(swipe.targets[,25]), 1])

##Look at the distribution of the e-values, alignment lengths and genome lengths
hist(-log10(swipe.targets$evalue.num), breaks = 200)
hist(swipe.targets$alignment.length.num, breaks = 200)
hist(swipe.targets$target.genome.length.num, breaks = 200)

##have a look at genomes with lengths recorded as less than 6000 nucleotides
length(unique(swipe.targets[swipe.targets[,25]< 500, 1]))

##count number of hist to target for each host genome
hits.count.num <- rep(NA, nrow(swipe.targets))
a <- 0
for(i in 1:50){
  a <- a + 1
  if(a == 1){
   a <- 0 
  }
  x <- swipe.targets[i,16]
  host.dat <- swipe.targets[swipe.targets[,16]==x,]
  y <- swipe.targets[i, 1]
  target.dat <- host.dat[host.dat[,1]==y,]
  hits.count.num[i] <- nrow(target.dat)
}
dat <- cbind(swipe.targets, hits.count.num)

```


```{r  protospacer_analysis_tables, results='asis'}

swipe.targets <- read.table("~/Desktop/Project/CRISPRClustering/refseq_79.swipe.targets_nr", comment.char = "", fill = T, sep = "\t", header = T)
swipe.targets <- transform.df(swipe.targets)

tab_a <- as.data.frame(table(swipe.targets$hits.count.num))
colnames(tab_a) <- c("Number of Hits", "Frequency")

tab_a <- xtable(tab_a, caption=('Frequency of the number of hits to each target genome'))
print(tab_a, type="latex", caption.placement='top', comment=FALSE)
```
